{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "ae510ffa-c83d-41bd-a5b0-b49e137960a5",
   "metadata": {},
   "source": [
    "## Setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "57758fe8-0b53-4dca-aebc-6e933e41ed3b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load required packages\n",
    "library(tidyverse)\n",
    "library(ggVennDiagram)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d2c89edc-ba93-4f91-9874-85f0ab3a067f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Load all data\n",
    "load_fst_data <- function(path){\n",
    "    name <- basename(path)\n",
    "    win_size <- str_extract(name, \"(?<=win)\\\\d+(?=\\\\.txt)\")\n",
    "    df <- read_delim(path, delim = \"\\t\") %>% \n",
    "        mutate(win_size = win_size)\n",
    "    return(df)\n",
    "}\n",
    "all_fst_df <- snakemake@input[[\"all_fsts\"]] %>% \n",
    "    purrr::map_dfr(load_fst_data) %>% \n",
    "    mutate(win_size = as.character(win_size))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0a69b386-dacc-4728-9641-ef838d9645d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_fst_df_filtered <- all_fst_df %>% \n",
    "    filter(!(arg_site_fst < -1 & win_size == 1)) %>% \n",
    "    filter(!(no_snps < 20 & win_size == 10000))\n",
    "all_fst_df_filtered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "20e88547-c85e-46af-ab52-61842c6eaec6",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 14, repr.plot.height = 6)\n",
    "labels <- c(`1` = \"Per-site\", `10000` = '10 kb windows')\n",
    "site_fst_winSize_hist <- all_fst_df %>% \n",
    "    ggplot(aes(x = arg_site_fst)) +\n",
    "        geom_histogram(bins = 100, color = \"black\", fill = \"red\") +\n",
    "        facet_wrap(~win_size, scales = \"free\", labeller = as_labeller(labels)) +\n",
    "        ylab(\"Number of sites\") + xlab(\"Site-based Fst from tskit (i.e. ARGs)\") +\n",
    "        theme_classic() +\n",
    "        theme(axis.text = element_text(size = 13),\n",
    "              axis.title = element_text(size = 15),\n",
    "              strip.text.x = element_text(size = 13))\n",
    "site_fst_winSize_hist\n",
    "\n",
    "ggsave(filename = snakemake@output[[\"site_fst_winSize_hist\"]], plot = site_fst_winSize_hist,\n",
    "       device = \"pdf\", width = 14, height = 6, units = \"in\", dpi = 300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "63abf391-79d8-43fc-be97-f7fdb4082a96",
   "metadata": {},
   "outputs": [],
   "source": [
    "site_fst_winSize_hist_filt <- all_fst_df_filtered %>% \n",
    "    ggplot(aes(x = arg_site_fst)) +\n",
    "        geom_histogram(bins = 100, color = \"black\", fill = \"red\") +\n",
    "        facet_wrap(~win_size, scales = \"free\", labeller = as_labeller(labels)) +\n",
    "        ylab(\"Number of sites\") + xlab(\"Site-based Fst from tskit (i.e. ARGs)\") +\n",
    "        theme_classic() +\n",
    "        theme(axis.text = element_text(size = 13),\n",
    "              axis.title = element_text(size = 15),\n",
    "              strip.text.x = element_text(size = 13))\n",
    "site_fst_winSize_hist_filt\n",
    "\n",
    "ggsave(filename = snakemake@output[[\"site_fst_winSize_hist_filt\"]], plot = site_fst_winSize_hist_filt,\n",
    "       device = \"pdf\", width = 14, height = 6, units = \"in\", dpi = 300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3b03d2c7-6bf1-48ed-b02f-0ba465159e4b",
   "metadata": {},
   "outputs": [],
   "source": [
    "branch_fst_winSize_hist_filt <- all_fst_df_filtered %>% \n",
    "    ggplot(aes(x = arg_branch_fst)) +\n",
    "        geom_histogram(bins = 100, color = \"black\", fill = \"red\") +\n",
    "        facet_wrap(~win_size, scales = \"free\", labeller = as_labeller(labels)) +\n",
    "        ylab(\"Number of sites\") + xlab(\"Branch-based Fst from tskit (i.e. ARGs)\") +\n",
    "        theme_classic() +\n",
    "        theme(axis.text = element_text(size = 13),\n",
    "              axis.title = element_text(size = 15),\n",
    "              strip.text.x = element_text(size = 13))\n",
    "branch_fst_winSize_hist_filt\n",
    "\n",
    "ggsave(filename = snakemake@output[[\"branch_fst_winSize_hist_filt\"]], plot = branch_fst_winSize_hist_filt,\n",
    "       device = \"pdf\", width = 14, height = 6, units = \"in\", dpi = 300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6940079b-26d6-41d0-8ecc-f892b2b648ef",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "options(repr.plot.width = 20, repr.plot.height = 12)\n",
    "site_fst_by_method_winSize_manhat <- all_fst_df_filtered %>% \n",
    "    ggplot() +\n",
    "        geom_line(aes(x = window_pos_1, y = gt_hudson_fst, color = \"Genotype (VCF)\"), linewidth = 0.5, alpha = 0.5) +\n",
    "        geom_line(aes(x = window_pos_1, y = arg_site_fst, color = \"ARG-site\"), linewidth = 0.5, alpha = 0.8) +\n",
    "        facet_grid(win_size~., scales = \"free\", labeller = as_labeller(labels)) +\n",
    "        ylab(\"Fst\") + xlab(\"Genomic position\") +\n",
    "        theme_classic() +\n",
    "        scale_color_manual(name = \"Method\", values = c(\"Genotype (VCF)\" = \"blue\", \"ARG-site\" = \"red\")) +\n",
    "        theme(axis.text = element_text(size = 13),\n",
    "              axis.title = element_text(size = 15),\n",
    "              strip.text.x = element_text(size = 13),\n",
    "              legend.position = \"top\",\n",
    "              legend.direction = \"horizontal\",\n",
    "              legend.key.size = unit(1, 'cm'),\n",
    "              legend.title = element_blank(),\n",
    "              legend.text = element_text(size=13))\n",
    "site_fst_by_method_winSize_manhat\n",
    "\n",
    "ggsave(filename = snakemake@output[[\"site_fst_by_method_winSize_manhat\"]], plot = site_fst_by_method_winSize_manhat,\n",
    "       device = \"pdf\", width = 20, height = 12, units = \"in\", dpi = 300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f49ec92f-63b7-4ecc-b735-a60e81e95e86",
   "metadata": {},
   "outputs": [],
   "source": [
    "branch_fst_by_method_winSize_manhat <- all_fst_df_filtered %>% \n",
    "    ggplot() +\n",
    "        geom_line(aes(x = window_pos_1, y = gt_hudson_fst, color = \"Genotype (VCF)\"), linewidth = 0.5, alpha = 0.5) +\n",
    "        geom_line(aes(x = window_pos_1, y = arg_branch_fst, color = \"ARG-branch\"), linewidth = 0.5) +\n",
    "        facet_grid(win_size~., scales = \"free\", labeller = as_labeller(labels)) +\n",
    "        ylab(\"Fst\") + xlab(\"Genomic position\") +\n",
    "        theme_classic() +\n",
    "        scale_color_manual(name = \"Method\", values = c(\"Genotype (VCF)\" = \"blue\", \"ARG-branch\" = \"red\")) +\n",
    "        theme(axis.text = element_text(size = 13),\n",
    "              axis.title = element_text(size = 15),\n",
    "              strip.text.x = element_text(size = 13),\n",
    "              legend.position = \"top\",\n",
    "              legend.direction = \"horizontal\",\n",
    "              legend.key.size = unit(1, 'cm'),\n",
    "              legend.title = element_blank(),\n",
    "              legend.text = element_text(size=13))\n",
    "branch_fst_by_method_winSize_manhat\n",
    "\n",
    "ggsave(filename = snakemake@output[[\"branch_fst_by_method_winSize_manhat\"]], plot = branch_fst_by_method_winSize_manhat,\n",
    "       device = \"pdf\",  width = 20, height = 12, units = \"in\", dpi = 300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3d69b050-2229-4d3b-8a34-6ba05222716b",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 14, repr.plot.height = 6)\n",
    "site_gt_fst_cor_by_winSize <- all_fst_df_filtered %>% \n",
    "    ggplot(aes(x = arg_site_fst, y = gt_hudson_fst)) +\n",
    "        geom_point(size = 2, alpha = 0.4) +\n",
    "        facet_wrap(~win_size, scales = \"free\", labeller = as_labeller(labels)) +\n",
    "        geom_smooth(method = \"lm\", linewidth = 1, color = 'blue') +\n",
    "        ylab(\"Hudson's Fst from Pixy (i.e. VCFs)\") + xlab(\"Site-based Fst from tskit (i.e. ARGs)\") +\n",
    "        theme_classic() +\n",
    "        theme(axis.text = element_text(size = 13),\n",
    "              axis.title = element_text(size = 15),\n",
    "              strip.text.x = element_text(size = 13))\n",
    "site_gt_fst_cor_by_winSize\n",
    "\n",
    "ggsave(filename = snakemake@output[[\"site_gt_fst_cor_by_winSize\"]], plot = site_gt_fst_cor_by_winSize,\n",
    "       device = \"pdf\", width = 14, height = 6, units = \"in\", dpi = 300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ce2c790e-5cab-4f75-87f7-2a894a37f060",
   "metadata": {},
   "outputs": [],
   "source": [
    "branch_gt_fst_cor_by_winSize <- all_fst_df_filtered %>% \n",
    "    ggplot(aes(x = arg_branch_fst, y = gt_hudson_fst)) +\n",
    "        geom_point(size = 2, alpha = 0.4) +\n",
    "        geom_smooth(method = \"lm\", linewidth = 1, color = 'blue') +\n",
    "        facet_wrap(~win_size, scales = \"free\", labeller = as_labeller(labels)) +\n",
    "        ylab(\"Hudson's Fst from Pixy (i.e. VCFs)\") + xlab(\"Branch-based Fst from tskit (i.e. ARGs)\") +\n",
    "        theme_classic() +\n",
    "        theme(axis.text = element_text(size = 13),\n",
    "              axis.title = element_text(size = 15),\n",
    "              strip.text.x = element_text(size = 13))\n",
    "branch_gt_fst_cor_by_winSize\n",
    "\n",
    "ggsave(filename = snakemake@output[[\"branch_gt_fst_cor_by_winSize\"]], plot = branch_gt_fst_cor_by_winSize,\n",
    "       device = \"pdf\", width = 14, height = 6, units = \"in\", dpi = 300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5fd733aa-eed5-4bde-8d75-5c1a7fd6f2e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "site_sfs_fst_cor_by_winSize <- all_fst_df_filtered %>% \n",
    "    ggplot(aes(x = arg_site_fst, y = sfs_hudson_fst)) +\n",
    "        geom_point(size = 2, alpha = 0.4) +\n",
    "        geom_smooth(method = \"lm\", linewidth = 1, color = 'blue') +\n",
    "        facet_wrap(~win_size, scales = \"free\", labeller = as_labeller(labels)) +\n",
    "        ylab(\"Hudson's Fst from ANGSD (i.e. VCFs)\") + xlab(\"Site-based Fst from tskit (i.e. ARGs)\") +\n",
    "        theme_classic() +\n",
    "        theme(axis.text = element_text(size = 13),\n",
    "              axis.title = element_text(size = 15),\n",
    "              strip.text.x = element_text(size = 13))\n",
    "site_sfs_fst_cor_by_winSize\n",
    "\n",
    "ggsave(filename = snakemake@output[[\"site_sfs_fst_cor_by_winSize\"]], plot = site_sfs_fst_cor_by_winSize,\n",
    "       device = \"pdf\", width = 14, height = 6, units = \"in\", dpi = 300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e7d071e5-513a-49f9-afeb-6218f8a57ef5",
   "metadata": {},
   "outputs": [],
   "source": [
    "branch_sfs_fst_cor_by_winSize <- all_fst_df_filtered %>% \n",
    "    ggplot(aes(x = arg_branch_fst, y = sfs_hudson_fst)) +\n",
    "        geom_point(size = 2, alpha = 0.4) +\n",
    "        geom_smooth(method = \"lm\", linewidth = 1, color = 'blue') +\n",
    "        facet_wrap(~win_size, scales = \"free\", labeller = as_labeller(labels)) +\n",
    "        ylab(\"Hudson's Fst from ANGSD (i.e. VCFs)\") + xlab(\"Branch-based Fst from tskit (i.e. ARGs)\") +\n",
    "        theme_classic() +\n",
    "        theme(axis.text = element_text(size = 13),\n",
    "              axis.title = element_text(size = 15),\n",
    "              strip.text.x = element_text(size = 13))\n",
    "branch_sfs_fst_cor_by_winSize\n",
    "\n",
    "ggsave(filename = snakemake@output[[\"branch_sfs_fst_cor_by_winSize\"]], plot = branch_sfs_fst_cor_by_winSize,\n",
    "       device = \"pdf\", width = 14, height = 6, units = \"in\", dpi = 300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4e9a75c0-4d55-44a2-9f95-7800211ee656",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create dataframe with Fst correlation coefficients\n",
    "all_fst_cors <- all_fst_df_filtered %>% \n",
    "    group_by(win_size, regionID) %>% \n",
    "    summarize(branch_gt_cor = cor(arg_branch_fst, gt_hudson_fst, use = \"complete.obs\"),\n",
    "              site_gt_cor = cor(arg_site_fst, gt_hudson_fst, use = \"complete.obs\"),\n",
    "              branch_sfs_cor = cor(arg_branch_fst, sfs_hudson_fst, use = \"complete.obs\"),\n",
    "              site_sfs_cor = cor(arg_site_fst, sfs_hudson_fst, use = \"complete.obs\"))\n",
    "all_fst_cors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "73f18eac-f78b-4056-8a09-0e30ac0f698b",
   "metadata": {},
   "outputs": [],
   "source": [
    "branch_gt_cor_hist_by_winSize <- all_fst_cors %>% \n",
    "    ggplot(aes(x = branch_gt_cor)) +\n",
    "        geom_histogram(bins = 30, color = \"black\", fill = \"red\") +\n",
    "        ylab(\"Number of 1Mb regions\") + xlab(\"ARG branch Fst vs. Pixy GT Fst correlation\") +\n",
    "        facet_wrap(~win_size, scales = \"free\", labeller = as_labeller(labels)) +\n",
    "        theme_classic() +\n",
    "        theme(axis.text = element_text(size = 13),\n",
    "              axis.title = element_text(size = 15),\n",
    "              strip.text.x = element_text(size = 13))\n",
    "branch_gt_cor_hist_by_winSize\n",
    "\n",
    "ggsave(filename = snakemake@output[[\"branch_gt_cor_hist_by_winSize\"]], plot = branch_gt_cor_hist_by_winSize,\n",
    "       device = \"pdf\", width = 14, height = 6, units = \"in\", dpi = 300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "734a7eb7-bcac-4537-985d-4776139c5257",
   "metadata": {},
   "outputs": [],
   "source": [
    "branch_gt_cor_hist_by_winSize <- all_fst_cors %>% \n",
    "    ggplot(aes(x = branch_gt_cor)) +\n",
    "        geom_histogram(bins = 30, color = \"black\", fill = \"red\") +\n",
    "        ylab(\"Number of 1Mb regions\") + xlab(\"ARG branch Fst vs. Pixy GT Fst correlation\") +\n",
    "        facet_wrap(~win_size, scales = \"free\", labeller = as_labeller(labels)) +\n",
    "        theme_classic() +\n",
    "        theme(axis.text = element_text(size = 13),\n",
    "              axis.title = element_text(size = 15),\n",
    "              strip.text.x = element_text(size = 13))\n",
    "branch_gt_cor_hist_by_winSize\n",
    "\n",
    "ggsave(filename = snakemake@output[[\"branch_gt_cor_hist_by_winSize\"]], plot = branch_gt_cor_hist_by_winSize,\n",
    "       device = \"pdf\", width = 14, height = 6, units = \"in\", dpi = 300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a1e06e33-285b-4e99-aff0-8c6dcb65c6ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "site_gt_cor_hist_by_winSize <- all_fst_cors %>% \n",
    "    ggplot(aes(x = site_gt_cor)) +\n",
    "        geom_histogram(bins = 30, color = \"black\", fill = \"red\") +\n",
    "        ylab(\"Number of 1Mb regions\") + xlab(\"ARG site Fst vs. Pixy GT Fst correlation\") +\n",
    "        facet_wrap(~win_size, scales = \"free\", labeller = as_labeller(labels)) +\n",
    "        theme_classic() +\n",
    "        theme(axis.text = element_text(size = 13),\n",
    "              axis.title = element_text(size = 15),\n",
    "              strip.text.x = element_text(size = 13))\n",
    "site_gt_cor_hist_by_winSize\n",
    "\n",
    "ggsave(filename = snakemake@output[[\"site_gt_cor_hist_by_winSize\"]], plot = site_gt_cor_hist_by_winSize,\n",
    "       device = \"pdf\", width = 14, height = 6, units = \"in\", dpi = 300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "41575754-100d-449b-84ef-4bb3abb55f0b",
   "metadata": {},
   "outputs": [],
   "source": [
    "branch_sfs_cor_hist_by_winSize <- all_fst_cors %>% \n",
    "    ggplot(aes(x = branch_sfs_cor)) +\n",
    "        geom_histogram(bins = 30, color = \"black\", fill = \"red\") +\n",
    "        ylab(\"Number of 1Mb regions\") + xlab(\"ARG branch Fst vs. ANGSD SFS Fst correlation\") +\n",
    "        facet_wrap(~win_size, scales = \"free\", labeller = as_labeller(labels)) +\n",
    "        theme_classic() +\n",
    "        theme(axis.text = element_text(size = 13),\n",
    "              axis.title = element_text(size = 15),\n",
    "              strip.text.x = element_text(size = 13))\n",
    "branch_sfs_cor_hist_by_winSize\n",
    "\n",
    "ggsave(filename = snakemake@output[[\"branch_sfs_cor_hist_by_winSize\"]], plot = branch_sfs_cor_hist_by_winSize,\n",
    "       device = \"pdf\", width = 14, height = 6, units = \"in\", dpi = 300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2b91799c-8be3-47cc-bb73-20cdc516784a",
   "metadata": {},
   "outputs": [],
   "source": [
    "site_sfs_cor_hist_by_winSize <- all_fst_cors %>% \n",
    "    ggplot(aes(x = site_sfs_cor)) +\n",
    "        geom_histogram(bins = 30, color = \"black\", fill = \"red\") +\n",
    "        ylab(\"Number of 1Mb regions\") + xlab(\"ARG branch Fst vs. ANGSD SFS Fst correlation\") +\n",
    "        facet_wrap(~win_size, scales = \"free\", labeller = as_labeller(labels)) +\n",
    "        theme_classic() +\n",
    "        theme(axis.text = element_text(size = 13),\n",
    "              axis.title = element_text(size = 15),\n",
    "              strip.text.x = element_text(size = 13))\n",
    "site_sfs_cor_hist_by_winSize\n",
    "\n",
    "ggsave(filename = snakemake@output[[\"site_sfs_cor_hist_by_winSize\"]], plot = site_sfs_cor_hist_by_winSize,\n",
    "       device = \"pdf\", width = 14, height = 6, units = \"in\", dpi = 300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "27380c46-8484-44e0-babf-4020a181b582",
   "metadata": {},
   "outputs": [],
   "source": [
    "windowed_fst_df_filtered <- all_fst_df_filtered %>% \n",
    "    filter(win_size == 10000) %>% \n",
    "    mutate(id = paste0(chromosome, \":\", window_pos_1, \"-\", window_pos_2)) %>% \n",
    "    mutate(win_center = window_pos_1 + 5000)\n",
    "arg_fst_quant <- quantile(windowed_fst_df_filtered %>% pull(arg_branch_fst), probs = 0.99); arg_fst_quant\n",
    "gt_fst_quant <- quantile(windowed_fst_df_filtered %>% pull(gt_hudson_fst), probs = 0.99); gt_fst_quant\n",
    "sfs_fst_quant <- quantile(windowed_fst_df_filtered %>% pull(sfs_hudson_fst), probs = 0.99); sfs_fst_quant"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e6b04af6-55ff-4d11-8527-03b197520e7c",
   "metadata": {},
   "outputs": [],
   "source": [
    "arg_outliers <- windowed_fst_df_filtered %>% \n",
    "    filter(arg_branch_fst >= arg_fst_quant) %>% \n",
    "    pull(id)\n",
    "gt_outliers <- windowed_fst_df_filtered %>% \n",
    "    filter(gt_hudson_fst >= gt_fst_quant) %>% \n",
    "    pull(id)\n",
    "sfs_outliers <- windowed_fst_df_filtered %>% \n",
    "    filter(sfs_hudson_fst >= sfs_fst_quant) %>% \n",
    "    pull(id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "446d4b9c-9d38-44f9-adf4-f36bc9f11870",
   "metadata": {},
   "outputs": [],
   "source": [
    "outliers_list <- list(ARG = arg_outliers, GT = gt_outliers, SFS = sfs_outliers)\n",
    "fst_outlier_venn <- ggVennDiagram(outliers_list)\n",
    "\n",
    "ggsave(filename = snakemake@output[[\"fst_outlier_venn\"]], plot = fst_outlier_venn,\n",
    "       device = \"pdf\", height = 8, width = 8, units = \"in\", dpi = 300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8d983668-3aee-422a-9e34-da67d39c0904",
   "metadata": {},
   "outputs": [],
   "source": [
    "# data_cum <- windowed_fst_df_filtered %>% \n",
    "#     group_by(chromosome) %>% \n",
    "#     summarise(max_WinCenter = max(win_center)) %>% \n",
    "#     mutate(WinCenter_add = lag(cumsum(max_WinCenter), default = 0)) %>% \n",
    "#     dplyr::select(chromosome, WinCenter_add)\n",
    "\n",
    "# windowed_fst_df_filtered_mod <- windowed_fst_df_filtered %>%\n",
    "#     inner_join(data_cum, by = \"chromosome\") %>% \n",
    "#     mutate(WinCenter_cum = win_center + WinCenter_add) \n",
    "\n",
    "# axis_set <- windowed_fst_df_filtered_mod %>% \n",
    "#   group_by(chromosome) %>% \n",
    "#   summarize(center = mean(WinCenter_cum))\n",
    "\n",
    "windowed_fst_df_filtered <- windowed_fst_df_filtered %>% \n",
    "    mutate(outlier_category = case_when(\n",
    "        id %in% arg_outliers & id %in% gt_outliers & id %in% sfs_outliers ~ \"Three-way outlier\",\n",
    "        id %in% arg_outliers & id %in% gt_outliers ~ \"Two-way outlier\",\n",
    "        id %in% arg_outliers & id %in% sfs_outliers ~ \"Two-way outlier\",\n",
    "        id %in% gt_outliers & id %in% sfs_outliers ~ \"Two-way outlier\",\n",
    "        id %in% gt_outliers ~ \"One-way outlier\",\n",
    "        id %in% arg_outliers ~ \"One-way outlier\",\n",
    "        id %in% sfs_outliers ~ \"One-way outlier\",\n",
    "        TRUE ~ \"Not outlier\"\n",
    "    ))\n",
    "\n",
    "outlier_manhat <- windowed_fst_df_filtered %>%\n",
    "    # mutate(chrom_cat = case_when(chromosome == 'Chr01_Occ' ~ 'One',\n",
    "    #                              chromosome == 'Chr01_Pall' ~ 'Two',\n",
    "    #                              chromosome == 'Chr02_Occ' ~ 'One',\n",
    "    #                              chromosome == 'Chr02_Pall' ~ 'Two',\n",
    "    #                              chromosome == 'Chr03_Occ' ~ 'One',\n",
    "    #                              chromosome == 'Chr03_Pall' ~ 'Two',\n",
    "    #                              chromosome == 'Chr04_Occ' ~ 'One',\n",
    "    #                              chromosome == 'Chr04_Pall' ~ 'Two',\n",
    "    #                              chromosome == 'Chr05_Occ' ~ 'One',\n",
    "    #                              chromosome == 'Chr05_Pall' ~ 'Two',\n",
    "    #                              chromosome == 'Chr06_Occ' ~ 'One',\n",
    "    #                              chromosome == 'Chr06_Pall' ~ 'Two',\n",
    "    #                              chromosome == 'Chr07_Occ' ~ 'One',\n",
    "    #                              chromosome == 'Chr07_Pall' ~ 'Two',\n",
    "    #                              chromosome == 'Chr08_Occ' ~ 'One',\n",
    "    #                              chromosome == 'Chr08_Pall' ~ 'Two')) %>%\n",
    "    ggplot(aes(x = window_pos_1, y = arg_branch_fst)) +\n",
    "        geom_point(shape = 21, size = 4, aes(fill = outlier_category, color = outlier_category, alpha = outlier_category)) +\n",
    "        geom_hline(yintercept = arg_fst_quant, color = \"grey40\", linetype = \"dashed\") +\n",
    "        scale_y_continuous(expand = c(0,0)) +\n",
    "        coord_cartesian(ylim = c(0, 0.05)) +\n",
    "        scale_fill_manual(values = c(\"black\", \"#003049\", \"#d62828\", \"#f77f00\")) + \n",
    "        scale_color_manual(values = c(\"black\", \"#003049\", \"#d62828\", \"#f77f00\")) + \n",
    "        scale_alpha_manual(values = c(0.4, 1, 1, 1, 1)) + \n",
    "        ylab('Branch-based Fst from tskit (i.e. ARGs)') + xlab('Genomic position') +\n",
    "        theme_classic() +\n",
    "        theme(axis.text = element_text(size = 13),\n",
    "              axis.title = element_text(size = 15),\n",
    "              strip.text.x = element_text(size = 13),\n",
    "              legend.position = \"top\",\n",
    "              legend.direction = \"horizontal\",\n",
    "              legend.key.size = unit(1, 'cm'),\n",
    "              legend.title = element_blank(),\n",
    "              legend.text = element_text(size=13))\n",
    "outlier_manhat\n",
    "\n",
    "ggsave(filename = snakemake@output[[\"outlier_manhat\"]], plot = outlier_manhat,\n",
    "       device = \"pdf\", width = 14, height = 6, units = \"in\", dpi = 300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d29d1a06-bbf5-482b-8dfe-d8c619a55b82",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.2.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
