{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "281fad93-79fb-44a7-832d-ea8510fb68a4",
   "metadata": {},
   "source": [
    "# Setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b15c49cf-aab5-4871-88bf-553b9d70d25a",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Load required packages\n",
    "library(tidyverse)\n",
    "library(bedtoolsr)\n",
    "library(vcfR)\n",
    "library(GenotypePlot)\n",
    "library(factoextra)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9bc8b197-555e-47e9-8408-2cfe9f60a03a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ggplot theme with transparent backgrounds\n",
    "transp_theme <- theme(panel.background = element_rect(fill = \"transparent\", colour = NA_character_), \n",
    "                      panel.grid.major = element_blank(),\n",
    "                      panel.grid.minor = element_blank(),\n",
    "                      plot.background = element_rect(fill = \"transparent\", colour = NA_character_),\n",
    "                      legend.background = element_rect(fill = \"transparent\"),\n",
    "                      legend.box.background = element_rect(fill = \"transparent\"),\n",
    "                      legend.key = element_rect(fill = \"transparent\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "27010823-5cea-490b-a4c0-2f960af5975e",
   "metadata": {},
   "source": [
    "# Analysis of selective sweeps \n",
    "\n",
    "## XP-nSL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "78ce5d4f-dfdc-4158-a8f6-072d171cde03",
   "metadata": {},
   "outputs": [],
   "source": [
    "remap_chr_names <- function(df){\n",
    "    df_out <- df %>% \n",
    "    mutate(Chr = case_when(Chr == 'Chr01_Occ' ~ 1,\n",
    "           Chr == 'Chr01_Pall' ~ 2,\n",
    "           Chr == 'Chr02_Occ' ~ 3,\n",
    "           Chr == 'Chr02_Pall' ~ 4,\n",
    "           Chr == 'Chr03_Occ' ~ 5,\n",
    "           Chr == 'Chr03_Pall' ~ 6,\n",
    "           Chr == 'Chr04_Occ' ~ 7,\n",
    "           Chr == 'Chr04_Pall' ~ 8,\n",
    "           Chr == 'Chr05_Occ' ~ 9,\n",
    "           Chr == 'Chr05_Pall' ~ 10,\n",
    "           Chr == 'Chr06_Occ' ~ 11,\n",
    "           Chr == 'Chr06_Pall' ~ 12,\n",
    "           Chr == 'Chr07_Occ' ~ 13,\n",
    "           Chr == 'Chr07_Pall' ~ 14,\n",
    "           Chr == 'Chr08_Occ' ~ 15,\n",
    "           Chr == 'Chr08_Pall' ~ 16))\n",
    "    return(df_out)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "01573d02-3201-4118-9553-9e279591a8f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load observed XP-nSL data\n",
    "obs_xpnsl_ur_df <- read_delim(snakemake@input[[\"win_xpnsl_ur\"]]) %>% \n",
    "    remap_chr_names()\n",
    "obs_xpnsl_ur_outRem_df <- read_delim(snakemake@input[[\"win_xpnsl_ur_outRem\"]]) %>% \n",
    "    remap_chr_names()\n",
    "obs_xpnsl_sr_df <- read_delim(snakemake@input[[\"win_xpnsl_sr\"]]) %>% \n",
    "    remap_chr_names()\n",
    "obs_xpnsl_us_df <- read_delim(snakemake@input[[\"win_xpnsl_us\"]]) %>% \n",
    "    remap_chr_names()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "58176512-7bf2-47df-b085-07b1a01d14f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Distribution of the number of sites per window\n",
    "xpnsl_nsites_thresh <- 50\n",
    "xpnsl_nSites_hist <- obs_xpnsl_ur_df %>% \n",
    "    ggplot(aes(x = n)) +\n",
    "        geom_histogram(bins = 50, color = \"black\", fill = \"grey\") +\n",
    "        ylab(\"Number of windows\") + xlab(\"Number of sites\") +\n",
    "        geom_vline(xintercept = xpnsl_nsites_thresh, color = \"red\", linetype = \"dashed\", linewidth = 1) +\n",
    "        theme_classic() +\n",
    "        theme(axis.text = element_text(size = 15),\n",
    "              axis.title = element_text(size = 17))\n",
    "\n",
    "xpnsl_nSites_hist\n",
    "ggsave(filename = snakemake@output[[\"xpnsl_nSites_hist\"]], plot = xpnsl_nSites_hist, \n",
    "       height = 8, width = 8, device = \"pdf\", dpi = 600, units = \"in\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bd38ec20-f7ae-4d9b-874c-c6aac81c777d",
   "metadata": {},
   "source": [
    "### Filter and identify outlier windows"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1f9130fb-b3d8-4d7c-9c8f-7e55f3ddd3db",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to assign outliers to XP-nSL windows\n",
    "assign_xpnsl_outliers <- function(df, comp){\n",
    "\n",
    "    if(comp == \"Urban_Rural\"){\n",
    "        pos_dir = \"Urban\"\n",
    "        neg_dir = \"Rural\"\n",
    "    } else if (comp == \"Suburban_Rural\"){\n",
    "        pos_dir = \"Suburban\"\n",
    "        neg_dir = \"Rural\"\n",
    "    } else if (comp == \"Urban_Suburban\"){\n",
    "        pos_dir = \"Urban\"\n",
    "        neg_dir = \"Suburban\"       \n",
    "    }\n",
    "\n",
    "    df_filt <- df %>%\n",
    "        filter(n >= xpnsl_nsites_thresh)\n",
    "\n",
    "    # Get critical values for mean XP-nSL score and proportions greater or lesser than 2 and -2, respectively\n",
    "    xpnsl_score_quant <- quantile(df_filt %>% pull(mean), probs = c(0.01, 0.99))\n",
    "    xpnsl_gtprop_quant <- quantile(df_filt %>% pull(gt_frac), probs = 0.99)\n",
    "    xpnsl_ltprop_quant <- quantile(df_filt %>% pull(lt_frac), probs = 0.99)\n",
    "\n",
    "    # Identify outliers and add as categorical variable to windows dataframe\n",
    "    df_filt <- df_filt %>%\n",
    "        mutate(xpnsl_score_outlier = ifelse(mean <= xpnsl_score_quant[1] | mean >= xpnsl_score_quant[2], 1, 0),\n",
    "               xpnsl_gtprop_outlier = ifelse(gt_frac >= xpnsl_gtprop_quant, 1, 0),\n",
    "               xpnsl_ltprop_outlier = ifelse(lt_frac >= xpnsl_ltprop_quant, 1, 0),\n",
    "               direction = case_when(xpnsl_score_outlier == 1 & mean > 0 & xpnsl_gtprop_outlier == 1 ~ pos_dir,\n",
    "                                     xpnsl_score_outlier == 1 & mean < 0 & xpnsl_ltprop_outlier == 1 ~ neg_dir,\n",
    "                                     TRUE ~ 'None')) %>% \n",
    "    mutate(prop_outlier = case_when(direction == pos_dir ~ gt_frac,\n",
    "                                    direction == neg_dir ~ lt_frac,\n",
    "                                    TRUE ~ NA))\n",
    "\n",
    "    return(df_filt)\n",
    "}\n",
    "\n",
    "obs_xpnsl_ur_df_filt <- assign_xpnsl_outliers(obs_xpnsl_ur_df, comp = \"Urban_Rural\")\n",
    "obs_xpnsl_ur_outRem_df_filt <- assign_xpnsl_outliers(obs_xpnsl_ur_outRem_df, comp = \"Urban_Rural\")\n",
    "obs_xpnsl_sr_df_filt <- assign_xpnsl_outliers(obs_xpnsl_sr_df, comp = \"Suburban_Rural\")\n",
    "obs_xpnsl_us_df_filt <- assign_xpnsl_outliers(obs_xpnsl_us_df, comp = \"Urban_Suburban\")\n",
    "\n",
    "sprintf(\"%s of %s XP-nSL windows remaining after removing those with less than %s sites for Urban-Rural comparison\", nrow(obs_xpnsl_ur_df_filt), nrow(obs_xpnsl_ur_df), xpnsl_nsites_thresh)\n",
    "sprintf(\"%s of %s XP-nSL windows remaining after removing those with less than %s sites for Urban-Rural (outlier removed) comparison\", nrow(obs_xpnsl_ur_outRem_df_filt), nrow(obs_xpnsl_ur_outRem_df), xpnsl_nsites_thresh)\n",
    "sprintf(\"%s of %s XP-nSL windows remaining after removing those with less than %s sites for Suburban-Rural comparison\", nrow(obs_xpnsl_sr_df_filt), nrow(obs_xpnsl_sr_df), xpnsl_nsites_thresh)\n",
    "sprintf(\"%s of %s XP-nSL windows remaining after removing those with less than %s sites for Urban-Suburban comparison\", nrow(obs_xpnsl_us_df_filt), nrow(obs_xpnsl_us_df), xpnsl_nsites_thresh)\n",
    "\n",
    "head(obs_xpnsl_ur_df_filt)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "60b1869d-5845-4ee7-a83a-46650c8f976b",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Add ranks and percentiles from permuted distributions"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1241ebcf-1df0-4a64-a52a-dab684d9345b",
   "metadata": {},
   "source": [
    "#### Correlation between observed and permuted XP-nSL across iterations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "34c1013d-5486-49e6-9c92-c9be79b2df03",
   "metadata": {},
   "outputs": [],
   "source": [
    "perm_xpnsl_df <- purrr::map(snakemake@input[[\"win_xpnsl_perm\"]], function(x) suppressMessages(read_delim(x, delim ='\\t'))) %>% \n",
    "    map_dfr(., assign_xpnsl_outliers, comp = \"Urban_Rural\") %>% \n",
    "    remap_chr_names()\n",
    "head(perm_xpnsl_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6178c34c-5763-4f03-a71f-97cafcaaacef",
   "metadata": {},
   "outputs": [],
   "source": [
    "calc_cor <- function(df){\n",
    "    val <- cor(df$mean, obs_xpnsl_ur_df_filt$mean, method = \"pearson\")\n",
    "    return(data.frame(cor = val))\n",
    "}\n",
    "cor_plot <- perm_xpnsl_df %>% \n",
    "    group_split(iter) %>% \n",
    "    purrr::map_dfr(., calc_cor) %>% \n",
    "    ggplot(aes(x = cor)) +\n",
    "        geom_histogram(bins = 50, color = \"black\", fill = \"grey\") +\n",
    "        ylab(\"Number of iterations\") + xlab(\"Permuted vs. observed XP-nSL Pearson's correlation coefficient\") +\n",
    "        theme_classic() +\n",
    "        theme(axis.title = element_text(size = 17),\n",
    "              axis.text = element_text(size = 15))\n",
    "\n",
    "options(repr.plot.width = 8, repr.plot.height = 8)\n",
    "cor_plot\n",
    "ggsave(filename = snakemake@output[[\"cor_plot\"]], plot = cor_plot, width = 8, \n",
    "       height = 8, units = \"in\", device = \"pdf\", dpi = 600)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "da0aeafc-a978-4f43-aca2-c2da1a7eafd6",
   "metadata": {},
   "source": [
    "#### Permutations and observed density plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "659b08d8-0388-4a46-b40b-a05a37a8e6f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "urban_sel_permuted <- perm_xpnsl_df %>% filter(direction == \"Urban\")\n",
    "urban_sel_observed <- obs_xpnsl_ur_df_filt %>% filter(direction == \"Urban\")\n",
    "\n",
    "urb_mean_plot <- ggplot() +\n",
    "    geom_density(data = urban_sel_permuted, aes(x = mean, fill = \"Permuted\", color = \"Permuted\"), alpha = 0.5) + \n",
    "    geom_density(data = urban_sel_observed, aes(x = mean, fill = \"Observed\", color = \"Observed\"), alpha = 0.5) +\n",
    "    ylab(\"Density\") + xlab(\"Mean windowed XP-nSL score\") +\n",
    "    scale_fill_manual(name = \"Distribution\", values = c(\"Permuted\" = \"#0070ec\", \"Observed\" = \"#003876\")) +\n",
    "    scale_color_manual(name = \"Distribution\", values = c(\"Permuted\" = \"#0070ec\", \"Observed\" = \"#003876\")) +\n",
    "    theme_classic() + \n",
    "    theme(axis.title = element_text(size = 17),\n",
    "          axis.text = element_text(size = 15)) +\n",
    "    transp_theme\n",
    "\n",
    "urb_mean_plot\n",
    "ggsave(filename = snakemake@output[[\"urb_mean_plot\"]], plot = urb_mean_plot, width = 8, \n",
    "       height = 8, units = \"in\", device = \"pdf\", dpi = 600)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a29b450e-e2c1-4ddf-bd37-d5fa3f2b3b22",
   "metadata": {},
   "outputs": [],
   "source": [
    "urb_prop_plot <- ggplot() +\n",
    "    geom_density(data = urban_sel_permuted, aes(x = prop_outlier, fill = \"Permuted\", color = \"Permuted\"), alpha = 0.5) + \n",
    "    geom_density(data = urban_sel_observed, aes(x = prop_outlier, fill = \"Observed\", color = \"Observed\"), alpha = 0.5) +\n",
    "    ylab(\"Density\") + xlab(\"Proportion of XP-nSL scores in window > 2\") +\n",
    "    scale_fill_manual(name = \"Distribution\", values = c(\"Permuted\" = \"#0070ec\", \"Observed\" = \"#003876\")) +\n",
    "    scale_color_manual(name = \"Distribution\", values = c(\"Permuted\" = \"#0070ec\", \"Observed\" = \"#003876\")) +\n",
    "    theme_classic() +\n",
    "    theme(axis.title = element_text(size = 17),\n",
    "          axis.text = element_text(size = 15)) +\n",
    "    transp_theme\n",
    "\n",
    "urb_prop_plot\n",
    "ggsave(filename = snakemake@output[[\"urb_prop_plot\"]], plot = urb_prop_plot, width = 8, \n",
    "       height = 8, units = \"in\", device = \"pdf\", dpi = 600)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "28181f84-780c-4021-9239-5caeb7706fca",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Dataframe with percentile in permuted distribution of each observed XP-nSL window\n",
    "urban_percentile <- urban_sel_observed %>% \n",
    "    rowwise() %>% \n",
    "    mutate(mean_perm_perc = sum(mean >= urban_sel_permuted$mean) / nrow(urban_sel_permuted),\n",
    "           prop_perm_perc = sum(prop_outlier >= urban_sel_permuted$prop_outlier) / nrow(urban_sel_permuted)) %>% \n",
    "    arrange(desc(mean_perm_perc)) %>% \n",
    "    dplyr::select(Chr, winID, start, end, winCenter, mean_perm_perc, prop_perm_perc)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2f2d8d80-e7c3-4b22-9657-06885ce243ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "rural_sel_permuted <- perm_xpnsl_df %>% filter(direction == \"Rural\")\n",
    "rural_sel_observed <- obs_xpnsl_ur_df_filt %>% filter(direction == \"Rural\")\n",
    "\n",
    "rur_mean_plot <- ggplot() +\n",
    "    geom_density(data = rural_sel_permuted, aes(x = mean, fill = \"Permuted\", color = \"Permuted\"), alpha = 0.5) + \n",
    "    geom_density(data = rural_sel_observed, aes(x = mean, fill = \"Observed\", color = \"Observed\"), alpha = 0.5) +\n",
    "    ylab(\"Density\") + xlab(\"Mean windowed XP-nSL score\") +\n",
    "    scale_fill_manual(name = \"Distribution\", values = c(\"Permuted\" = \"#00e989\", \"Observed\" = \"#007243\")) +\n",
    "    scale_color_manual(name = \"Distribution\", values = c(\"Permuted\" = \"#00e989\", \"Observed\" = \"#007243\")) +\n",
    "    theme_classic() +\n",
    "    theme(axis.title = element_text(size = 17),\n",
    "          axis.text = element_text(size = 15)) +\n",
    "    transp_theme\n",
    "\n",
    "rur_mean_plot\n",
    "ggsave(filename = snakemake@output[[\"rur_mean_plot\"]], plot = rur_mean_plot, width = 8, \n",
    "       height = 8, units = \"in\", device = \"pdf\", dpi = 600)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "33f9381f-7611-40d5-8f1f-1710c3a8fb61",
   "metadata": {},
   "outputs": [],
   "source": [
    "rur_prop_plot <- ggplot() +\n",
    "    geom_density(data = rural_sel_permuted, aes(x = prop_outlier, fill = \"Permuted\", color = \"Permuted\"), alpha = 0.5) + \n",
    "    geom_density(data = rural_sel_observed, aes(x = prop_outlier, fill = \"Observed\", color = \"Observed\"), alpha = 0.5) +\n",
    "    ylab(\"Density\") + xlab(\"Proportion of XP-nSL scores < -2\") +\n",
    "    scale_fill_manual(name = \"Distribution\", values = c(\"Permuted\" = \"#00e989\", \"Observed\" = \"#007243\")) +\n",
    "    scale_color_manual(name = \"Distribution\", values = c(\"Permuted\" = \"#00e989\", \"Observed\" = \"#007243\")) +\n",
    "    theme_classic() +\n",
    "    theme(axis.title = element_text(size = 17),\n",
    "          axis.text = element_text(size = 15)) +\n",
    "    transp_theme\n",
    "\n",
    "rur_prop_plot\n",
    "ggsave(filename = snakemake@output[[\"rur_prop_plot\"]], plot = rur_prop_plot, width = 8, \n",
    "       height = 8, units = \"in\", device = \"pdf\", dpi = 600)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7a37feb8-6d89-4f05-aa17-a6bf8bcc82cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Dataframe with percentile in permuted distribution of each observed XP-nSL window\n",
    "rural_percentile <- rural_sel_observed %>% \n",
    "    rowwise() %>% \n",
    "    mutate(mean_perm_perc = sum(mean <= rural_sel_permuted$mean) / nrow(rural_sel_permuted),\n",
    "           prop_perm_perc = sum(prop_outlier >= rural_sel_permuted$prop_outlier) / nrow(rural_sel_permuted)) %>% \n",
    "    arrange(desc(mean_perm_perc)) %>% \n",
    "    dplyr::select(Chr, winID, start, end, winCenter, mean_perm_perc, prop_perm_perc)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1e55af47-88e0-427c-a0b5-3d99633f154a",
   "metadata": {},
   "outputs": [],
   "source": [
    "obs_xpnsl_ur_df_filt <- obs_xpnsl_ur_df_filt %>% \n",
    "    left_join(., bind_rows(urban_percentile, rural_percentile), by = c(\"Chr\", \"winID\", \"start\", \"end\", \"winCenter\"))\n",
    "head(obs_xpnsl_ur_df_filt)\n",
    "\n",
    "write_delim(obs_xpnsl_ur_df_filt, snakemake@output[[\"xpnsl_df\"]], delim='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d477c07a-9504-470c-847b-c5bdfe6f90d5",
   "metadata": {},
   "source": [
    "#### Add ranks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e133e755-40c8-4fbd-87c6-5b855e62bf9c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to add ranks to XP-nSL windows based on the proportion of outlier scores\n",
    "add_xpnsl_ranks <- function(df){\n",
    "        \n",
    "    df_out <- df %>% \n",
    "        arrange(desc(prop_perm_perc)) %>% \n",
    "        mutate(xpnsl_rank = 1:n())\n",
    "    return(df_out)\n",
    "}\n",
    "\n",
    "obs_xpnsl_ur_df_filt <- obs_xpnsl_ur_df_filt %>% \n",
    "    group_split(direction) %>% \n",
    "    purrr::map_dfr(., add_xpnsl_ranks) %>% \n",
    "    mutate(xpnsl_rank = ifelse(direction == 'None', NA, xpnsl_rank))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3dcdcf73-092a-4cfc-86eb-d7a691d34492",
   "metadata": {},
   "source": [
    "### Generate Manhattan plot"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c806b70e-1934-40c2-8832-c57234c5bc48",
   "metadata": {},
   "source": [
    "#### Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "37b81e5f-1c23-4476-b476-262303808ebd",
   "metadata": {},
   "outputs": [],
   "source": [
    "create_xpnsl_manhat_ur <- function(df, comp){\n",
    "\n",
    "    pos_dir <- \"Urban\"\n",
    "    neg_dir <- \"Rural\"\n",
    "    pos_col <- \"#003876\"\n",
    "    neg_col <- \"#007243\"\n",
    "\n",
    "    # Setting up cummulative genome-wide x-axis\n",
    "    data_cum <- df %>% \n",
    "        group_by(Chr) %>% \n",
    "        summarise(max_winCenter = max(winCenter)) %>% \n",
    "        mutate(winCenter_add = lag(cumsum(max_winCenter), default = 0)) %>% \n",
    "        dplyr::select(Chr, winCenter_add)\n",
    "    \n",
    "    df_mod <- df %>%\n",
    "        inner_join(data_cum, by = \"Chr\") %>% \n",
    "        mutate(winCenter_cum = winCenter + winCenter_add) \n",
    "    \n",
    "    axis_set <- df_mod %>% \n",
    "      group_by(Chr) %>% \n",
    "      summarize(center = mean(winCenter_cum))\n",
    "    \n",
    "    # Get XP-nSL outliers for plotting as separate layers\n",
    "    urban_xpnsl_topPerm <- df_mod %>% filter(direction == pos_dir & mean_perm_perc >= 0.95 & prop_perm_perc >= 0.95)\n",
    "    rural_xpnsl_topPerm <- df_mod %>% filter(direction == neg_dir & mean_perm_perc >= 0.95 & prop_perm_perc >= 0.95)\n",
    "    xpnsl_not_outlier <- df_mod %>% filter(direction == 'None')\n",
    "    urban_xpnsl_outliers <- df_mod %>% filter(direction == pos_dir & mean_perm_perc < 0.95 & prop_perm_perc < 0.95)\n",
    "    rural_xpnsl_outliers <- df_mod %>% filter(direction == neg_dir & mean_perm_perc < 0.95 & prop_perm_perc < 0.95)\n",
    "    \n",
    "    # Get XP-nSL score quantile for plotting\n",
    "    xpnsl_score_quant <- quantile(df_mod %>% pull(mean), probs = c(0.01, 0.99))\n",
    "    \n",
    "    xpnsl_not_outlier <- xpnsl_not_outlier %>%\n",
    "        mutate(chrom_cat = case_when(Chr == '1' ~ 'One',\n",
    "                                     Chr == '2' ~ 'Two',\n",
    "                                     Chr == '3' ~ 'One',\n",
    "                                     Chr == '4' ~ 'Two',\n",
    "                                     Chr == '5' ~ 'One',\n",
    "                                     Chr == '6' ~ 'Two',\n",
    "                                     Chr == '7' ~ 'One',\n",
    "                                     Chr == '8' ~ 'Two',\n",
    "                                     Chr == '9' ~ 'One',\n",
    "                                     Chr == '10' ~ 'Two',\n",
    "                                     Chr == '11' ~ 'One',\n",
    "                                     Chr == '12' ~ 'Two',\n",
    "                                     Chr == '13' ~ 'One',\n",
    "                                     Chr == '14' ~ 'Two',\n",
    "                                     Chr == '15' ~ 'One',\n",
    "                                     Chr == '16' ~ 'Two'))\n",
    "    \n",
    "    manhat <- ggplot() +\n",
    "            geom_point(data = xpnsl_not_outlier, shape = 21, alpha = 0.4, size = 3, \n",
    "                       aes(x = winCenter_cum, y = mean, fill = chrom_cat, color = chrom_cat)) +\n",
    "            geom_point(data = urban_xpnsl_outliers, shape = 21, alpha = 1, size = 3, color = pos_col, fill = pos_col,\n",
    "                       aes(x = winCenter_cum, y = mean)) +\n",
    "            geom_point(data = rural_xpnsl_outliers, shape = 21, alpha = 1, size = 3, color = neg_col, fill = neg_col,\n",
    "                       aes(x = winCenter_cum, y = mean)) +\n",
    "            geom_point(data = urban_xpnsl_topPerm, shape = 23, alpha = 1, size = 4.5, color = pos_col, fill = pos_col,\n",
    "                       aes(x = winCenter_cum, y = mean)) +\n",
    "            geom_point(data = rural_xpnsl_topPerm, shape = 23, alpha = 1, size = 4.5, color = neg_col, fill = neg_col,\n",
    "                       aes(x = winCenter_cum, y = mean)) +\n",
    "            geom_hline(yintercept = xpnsl_score_quant, color = \"grey40\", linetype = \"dashed\") +\n",
    "            scale_x_continuous(label = axis_set$Chr, breaks = axis_set$center) +\n",
    "            scale_y_continuous(expand = c(0,0)) +\n",
    "            coord_cartesian(ylim = c(-6, 5)) +\n",
    "            scale_fill_manual(values = c(\"black\", \"grey40\")) + \n",
    "            scale_color_manual(values = c(\"black\", \"grey40\")) + \n",
    "            ylab('Urban-rural normalized XP-nSL') + xlab('Chromosomes') +\n",
    "            theme_classic() +\n",
    "            theme(\n",
    "                legend.position = \"none\",\n",
    "                panel.border = element_blank(),\n",
    "                panel.grid.major.x = element_blank(),\n",
    "                panel.grid.minor.x = element_blank(),\n",
    "                axis.text = element_text(size=16),\n",
    "                axis.title = element_text(size = 20),\n",
    "              ) +\n",
    "            transp_theme\n",
    "    \n",
    "    return(manhat)\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "532d26f3-7aca-4454-8b7c-38a4d1fa03d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "create_xpnsl_manhat_ur_outRem <- function(df, comp){\n",
    "\n",
    "    pos_dir <- \"Urban\"\n",
    "    neg_dir <- \"Rural\"\n",
    "    pos_col <- \"#003876\"\n",
    "    neg_col <- \"#007243\"\n",
    "\n",
    "    # Setting up cummulative genome-wide x-axis\n",
    "    data_cum <- df %>% \n",
    "        group_by(Chr) %>% \n",
    "        summarise(max_winCenter = max(winCenter)) %>% \n",
    "        mutate(winCenter_add = lag(cumsum(max_winCenter), default = 0)) %>% \n",
    "        dplyr::select(Chr, winCenter_add)\n",
    "    \n",
    "    df_mod <- df %>%\n",
    "        inner_join(data_cum, by = \"Chr\") %>% \n",
    "        mutate(winCenter_cum = winCenter + winCenter_add) \n",
    "    \n",
    "    axis_set <- df_mod %>% \n",
    "      group_by(Chr) %>% \n",
    "      summarize(center = mean(winCenter_cum))\n",
    "    \n",
    "    # Get XP-nSL outliers for plotting as separate layers\n",
    "    xpnsl_not_outlier <- df_mod %>% filter(direction == 'None')\n",
    "    urban_xpnsl_outliers <- df_mod %>% filter(direction == pos_dir)\n",
    "    rural_xpnsl_outliers <- df_mod %>% filter(direction == neg_dir)\n",
    "    \n",
    "    # Get XP-nSL score quantile for plotting\n",
    "    xpnsl_score_quant <- quantile(df_mod %>% pull(mean), probs = c(0.01, 0.99))\n",
    "    \n",
    "    xpnsl_not_outlier <- xpnsl_not_outlier %>%\n",
    "        mutate(chrom_cat = case_when(Chr == '1' ~ 'One',\n",
    "                                     Chr == '2' ~ 'Two',\n",
    "                                     Chr == '3' ~ 'One',\n",
    "                                     Chr == '4' ~ 'Two',\n",
    "                                     Chr == '5' ~ 'One',\n",
    "                                     Chr == '6' ~ 'Two',\n",
    "                                     Chr == '7' ~ 'One',\n",
    "                                     Chr == '8' ~ 'Two',\n",
    "                                     Chr == '9' ~ 'One',\n",
    "                                     Chr == '10' ~ 'Two',\n",
    "                                     Chr == '11' ~ 'One',\n",
    "                                     Chr == '12' ~ 'Two',\n",
    "                                     Chr == '13' ~ 'One',\n",
    "                                     Chr == '14' ~ 'Two',\n",
    "                                     Chr == '15' ~ 'One',\n",
    "                                     Chr == '16' ~ 'Two'))\n",
    "    \n",
    "    manhat <- ggplot() +\n",
    "            geom_point(data = xpnsl_not_outlier, shape = 21, alpha = 0.4, size = 3, \n",
    "                       aes(x = winCenter_cum, y = mean, fill = chrom_cat, color = chrom_cat)) +\n",
    "            geom_point(data = urban_xpnsl_outliers, shape = 21, alpha = 1, size = 3, color = pos_col, fill = pos_col,\n",
    "                       aes(x = winCenter_cum, y = mean)) +\n",
    "            geom_point(data = rural_xpnsl_outliers, shape = 21, alpha = 1, size = 3, color = neg_col, fill = neg_col,\n",
    "                       aes(x = winCenter_cum, y = mean)) +\n",
    "            geom_hline(yintercept = xpnsl_score_quant, color = \"grey40\", linetype = \"dashed\") +\n",
    "            scale_x_continuous(label = axis_set$Chr, breaks = axis_set$center) +\n",
    "            scale_y_continuous(expand = c(0,0)) +\n",
    "            coord_cartesian(ylim = c(-6, 5)) +\n",
    "            scale_fill_manual(values = c(\"black\", \"grey40\")) + \n",
    "            scale_color_manual(values = c(\"black\", \"grey40\")) + \n",
    "            ylab('Urban-rural normalized XP-nSL') + xlab('Chromosomes') +\n",
    "            theme_classic() +\n",
    "            theme(\n",
    "                legend.position = \"none\",\n",
    "                panel.border = element_blank(),\n",
    "                panel.grid.major.x = element_blank(),\n",
    "                panel.grid.minor.x = element_blank(),\n",
    "                axis.text = element_text(size=16),\n",
    "                axis.title = element_text(size = 20),\n",
    "              ) +\n",
    "            transp_theme\n",
    "    \n",
    "    return(manhat)\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ffb313ac-6c69-42e0-a1cb-b72a40d5bbc3",
   "metadata": {},
   "outputs": [],
   "source": [
    "create_xpnsl_manhat_sr_us <- function(df, comp){\n",
    "\n",
    "    if(comp == \"Suburban_Rural\"){\n",
    "        pos_dir = \"Suburban\"\n",
    "        neg_dir = \"Rural\"\n",
    "        pos_col <- \"#914205\"\n",
    "        neg_col <- \"#007243\"\n",
    "        ylab <- \"Suburban-rural normalized XP-nSL\"\n",
    "    } else if (comp == \"Urban_Suburban\"){\n",
    "        pos_dir = \"Urban\"\n",
    "        neg_dir = \"Suburban\"  \n",
    "        pos_col <- \"#003876\"\n",
    "        neg_col <- \"#914205\"\n",
    "        ylab <- \"Urban-suburban normalized XP-nSL\"\n",
    "    }\n",
    "    \n",
    "    # Setting up cummulative genome-wide x-axis\n",
    "    data_cum <- df %>% \n",
    "        group_by(Chr) %>% \n",
    "        summarise(max_winCenter = max(winCenter)) %>% \n",
    "        mutate(winCenter_add = lag(cumsum(max_winCenter), default = 0)) %>% \n",
    "        dplyr::select(Chr, winCenter_add)\n",
    "    \n",
    "    df_mod <- df %>%\n",
    "        inner_join(data_cum, by = \"Chr\") %>% \n",
    "        mutate(winCenter_cum = winCenter + winCenter_add) \n",
    "    \n",
    "    axis_set <- df_mod %>% \n",
    "      group_by(Chr) %>% \n",
    "      summarize(center = mean(winCenter_cum))\n",
    "    \n",
    "    # Get XP-nSL outliers for plotting as separate layers\n",
    "    xpnsl_not_outlier <- df_mod %>% filter(direction == 'None')\n",
    "    urban_xpnsl_outliers <- df_mod %>% filter(direction == pos_dir)\n",
    "    rural_xpnsl_outliers <- df_mod %>% filter(direction == neg_dir)\n",
    "    \n",
    "    # Get XP-nSL score quantile for plotting\n",
    "    xpnsl_score_quant <- quantile(df_mod %>% pull(mean), probs = c(0.01, 0.99))\n",
    "    \n",
    "    xpnsl_not_outlier <- xpnsl_not_outlier %>%\n",
    "        mutate(chrom_cat = case_when(Chr == 1 ~ 'One',\n",
    "                                     Chr == 2 ~ 'Two',\n",
    "                                     Chr == 3 ~ 'One',\n",
    "                                     Chr == 4 ~ 'Two',\n",
    "                                     Chr == 5 ~ 'One',\n",
    "                                     Chr == 6 ~ 'Two',\n",
    "                                     Chr == 7 ~ 'One',\n",
    "                                     Chr == 8 ~ 'Two',\n",
    "                                     Chr == 9 ~ 'One',\n",
    "                                     Chr == 10 ~ 'Two',\n",
    "                                     Chr == 11 ~ 'One',\n",
    "                                     Chr == 12 ~ 'Two',\n",
    "                                     Chr == 13 ~ 'One',\n",
    "                                     Chr == 14 ~ 'Two',\n",
    "                                     Chr == 15 ~ 'One',\n",
    "                                     Chr == 16 ~ 'Two'))\n",
    "    \n",
    "    manhat <- ggplot() +\n",
    "            geom_point(data = xpnsl_not_outlier, shape = 21, alpha = 0.4, size = 3, \n",
    "                       aes(x = winCenter_cum, y = mean, fill = chrom_cat, color = chrom_cat)) +\n",
    "            geom_point(data = urban_xpnsl_outliers, shape = 21, alpha = 1, size = 3, color = pos_col, fill = pos_col,\n",
    "                       aes(x = winCenter_cum, y = mean)) +\n",
    "            geom_point(data = rural_xpnsl_outliers, shape = 21, alpha = 1, size = 3, color = neg_col, fill = neg_col,\n",
    "                       aes(x = winCenter_cum, y = mean)) +\n",
    "            geom_hline(yintercept = xpnsl_score_quant, color = \"grey40\", linetype = \"dashed\") +\n",
    "            scale_x_continuous(label = axis_set$Chr, breaks = axis_set$center) +\n",
    "            scale_y_continuous(expand = c(0,0)) +\n",
    "            coord_cartesian(ylim = c(-6, 5)) +\n",
    "            scale_fill_manual(values = c(\"black\", \"grey40\")) + \n",
    "            scale_color_manual(values = c(\"black\", \"grey40\")) + \n",
    "            ylab(ylab) + xlab('Chromosomes') +\n",
    "            theme_classic() +\n",
    "            theme(\n",
    "                legend.position = \"none\",\n",
    "                panel.border = element_blank(),\n",
    "                panel.grid.major.x = element_blank(),\n",
    "                panel.grid.minor.x = element_blank(),\n",
    "                axis.text = element_text(size=16),\n",
    "                axis.title = element_text(size = 20)\n",
    "              ) +\n",
    "            transp_theme\n",
    "    \n",
    "    return(manhat)\n",
    "}\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "27182820-1667-4552-8119-e53228964f62",
   "metadata": {},
   "source": [
    "#### Generate plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ffec3ad4-e8b0-4400-856a-93832bc11aae",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 20, repr.plot.height = 6)\n",
    "xpnsl_manhat_ur <- create_xpnsl_manhat_ur(obs_xpnsl_ur_df_filt, comp = \"Urban_Rural\")\n",
    "xpnsl_manhat_ur\n",
    "ggsave(filename = snakemake@output[[\"xpnsl_manhat_ur\"]], plot = xpnsl_manhat_ur, \n",
    "       height = 8, width = 20, device = \"pdf\", dpi = 600, units = \"in\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3a5f3221-801b-4e8c-ad14-5ab9787cafec",
   "metadata": {},
   "outputs": [],
   "source": [
    "xpnsl_manhat_ur_outRem <- create_xpnsl_manhat_ur_outRem(obs_xpnsl_ur_outRem_df_filt, comp = \"Urban_Rural\")\n",
    "xpnsl_manhat_ur_outRem\n",
    "ggsave(filename = snakemake@output[[\"xpnsl_manhat_ur_outRem\"]], plot = xpnsl_manhat_ur_outRem, \n",
    "       height = 8, width = 20, device = \"pdf\", dpi = 600, units = \"in\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6e322ea9-4bc7-4c70-b887-ad865b72e69a",
   "metadata": {},
   "outputs": [],
   "source": [
    "xpnsl_manhat_sr <- create_xpnsl_manhat_sr_us(obs_xpnsl_sr_df_filt, comp = \"Suburban_Rural\")\n",
    "xpnsl_manhat_sr\n",
    "ggsave(filename = snakemake@output[[\"xpnsl_manhat_sr\"]], plot = xpnsl_manhat_sr, \n",
    "       height = 8, width = 20, device = \"pdf\", dpi = 600, units = \"in\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f576cb22-30d5-4270-b663-4cfc387e3317",
   "metadata": {},
   "outputs": [],
   "source": [
    "xpnsl_manhat_us <- create_xpnsl_manhat_sr_us(obs_xpnsl_us_df_filt, comp = \"Urban_Suburban\")\n",
    "xpnsl_manhat_us\n",
    "ggsave(filename = snakemake@output[[\"xpnsl_manhat_us\"]], plot = xpnsl_manhat_us, \n",
    "       height = 8, width = 20, device = \"pdf\", dpi = 600, units = \"in\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f8433b9d-08aa-4316-840d-7cb0af61b73c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Suburban outliers shared with Urban vs. Rural\n",
    "ur_urban_ids <- obs_xpnsl_ur_df_filt %>% filter(direction == \"Urban\") %>% mutate(id = paste0(Chr, \"_\", winCenter)) %>% pull(id)\n",
    "ur_rural_ids <- obs_xpnsl_ur_df_filt %>% filter(direction == \"Rural\") %>% mutate(id = paste0(Chr, \"_\", winCenter)) %>% pull(id)\n",
    "sr_suburban_ids <- obs_xpnsl_sr_df_filt %>% filter(direction == \"Suburban\") %>% mutate(id = paste0(Chr, \"_\", winCenter)) %>% pull(id)\n",
    "\n",
    "sum(ur_urban_ids %in% sr_suburban_ids) / length(ur_urban_ids)\n",
    "sum(ur_rural_ids %in% sr_suburban_ids) / length(ur_rural_ids)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5a099a09-d2cf-41df-881e-3dde3852a873",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Overlap in urban and rural outliers with an without removal out population structure outlier populations\n",
    "ur_urban_ids <- obs_xpnsl_ur_df_filt %>% filter(direction == \"Urban\") %>% mutate(id = paste0(Chr, \"_\", winCenter)) %>% pull(id)\n",
    "ur_urban_top_ids <- obs_xpnsl_ur_df_filt %>% \n",
    "    filter(direction == \"Urban\" & mean_perm_perc >= 0.95 & prop_perm_perc >= 0.95) %>% \n",
    "    mutate(id = paste0(Chr, \"_\", winCenter)) %>% pull(id)\n",
    "ur_urban_outRem_ids <- obs_xpnsl_ur_outRem_df_filt %>% filter(direction == \"Urban\") %>% mutate(id = paste0(Chr, \"_\", winCenter)) %>% pull(id)\n",
    "\n",
    "\n",
    "sum(ur_urban_ids %in% ur_urban_outRem_ids) / length(ur_urban_ids)\n",
    "sum(ur_urban_top_ids %in% ur_urban_outRem_ids) / length(ur_urban_top_ids)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fc9e6b2e-e720-4724-90d1-b80e5d163a48",
   "metadata": {},
   "outputs": [],
   "source": [
    "ur_rural_ids <- obs_xpnsl_ur_df_filt %>% filter(direction == \"Rural\") %>% mutate(id = paste0(Chr, \"_\", winCenter)) %>% pull(id)\n",
    "ur_rural_top_ids <- obs_xpnsl_ur_df_filt %>% \n",
    "    filter(direction == \"Rural\" & mean_perm_perc >= 0.95 & prop_perm_perc >= 0.95) %>% \n",
    "    mutate(id = paste0(Chr, \"_\", winCenter)) %>% pull(id)\n",
    "ur_rural_outRem_ids <- obs_xpnsl_ur_outRem_df_filt %>% filter(direction == \"Rural\") %>% mutate(id = paste0(Chr, \"_\", winCenter)) %>% pull(id)\n",
    "\n",
    "\n",
    "sum(ur_rural_ids %in% ur_rural_outRem_ids) / length(ur_rural_ids)\n",
    "sum(ur_rural_top_ids %in% ur_rural_outRem_ids) / length(ur_rural_top_ids)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b3c6e498-1296-4cde-8107-4d244928e19e",
   "metadata": {},
   "source": [
    "### Merge XP-nSL outlier windows and get top 10"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "69c5538d-cb52-4b4c-aa3c-00bdba6ebc5d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to merge consecutive outlier windows\n",
    "merge_xpnsl_windows <- function(df, win_dist = 0){\n",
    "\n",
    "    dir <- df %>% pull(direction) %>% unique()\n",
    "    df_sorted <- df %>% \n",
    "        dplyr::select(Chr, start, end, everything()) %>% \n",
    "        arrange(Chr, start) %>% \n",
    "        mutate(prop_outlier = round(prop_outlier, 3),\n",
    "               mean_perm_perc = round(mean_perm_perc, 3),\n",
    "               prop_perm_perc = round(prop_perm_perc, 3)) %>% \n",
    "        mutate(is_topPerm = ifelse(mean_perm_perc >= 0.95 & prop_perm_perc >= 0.95, 1, 0))\n",
    "\n",
    "    col_names <- c('Chr', 'start', 'end', 'mean_xpnsl', 'min_max', 'direction', 'mean_prop_outlier', 'all_prop_outlier', 'mean_percentile_permuted', 'all_percentile_permuted','mean_prop_percentile_permuted','all_prop_percentile_permuted', 'min_xpnsl_rank', 'all_xpnsl_ranks', 'num_topPerm')\n",
    "    if(dir == 'Urban'){\n",
    "        cols <- c('6,7,17,18,18,19,19,20,20,21,21,22')\n",
    "        operation <- c('mean,max,distinct,mean,collapse,mean,collapse,mean,collapse,min,collapse,sum')\n",
    "    }else if(dir == 'Rural'){\n",
    "        cols <- c('6,8,17,18,18,19,19,20,20,21,21,22')\n",
    "        operation <- c('mean,min,distinct,mean,collapse,mean,collapse,mean,collapse,min,collapse,sum')\n",
    "    }\n",
    "    \n",
    "    df_merged  <- bt.merge(i = df_sorted, c = cols, o = operation, d = win_dist)\n",
    "    names(df_merged) <- col_names\n",
    "    df_merged <- df_merged %>% \n",
    "        mutate(win_size = end - start)\n",
    "\n",
    "    print(sprintf('%s: There were %s XP-nSL outlier windows prior to merging. There are %s outlier regions after merging consecutive outlier windows', dir, nrow(df_sorted), nrow(df_merged)))\n",
    "    return(df_merged)\n",
    "}\n",
    "\n",
    "win_xpnsl_ur_df_filt_merged <- obs_xpnsl_ur_df_filt %>% \n",
    "    filter(direction != 'None') %>% \n",
    "    group_split(direction) %>% \n",
    "    purrr::map_dfr(., merge_xpnsl_windows)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d5719b2f-2199-481e-a91e-bc9ed71fc900",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "xpnsl_topHits_urban_rural <- win_xpnsl_ur_df_filt_merged %>% \n",
    "    filter(num_topPerm > 0) %>% \n",
    "    group_by(direction) %>% \n",
    "    arrange(min_xpnsl_rank, .by_group = TRUE) %>% \n",
    "    dplyr::select(Chr, start, end, win_size, direction, mean_xpnsl, min_max, all_prop_outlier, mean_prop_outlier, mean_percentile_permuted,all_percentile_permuted, mean_prop_percentile_permuted, all_prop_percentile_permuted, all_xpnsl_ranks, min_xpnsl_rank, num_topPerm) %>% \n",
    "    ungroup()\n",
    "xpnsl_topHits_urban_rural"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f68c2cc8-f78f-4fca-952a-2b03ec3135b6",
   "metadata": {},
   "source": [
    "## saltiLassi"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f2517b60-8598-4306-8e62-0b440bf9f995",
   "metadata": {},
   "outputs": [],
   "source": [
    "urban_lassi_df <- read_delim(snakemake@input[[\"lassip\"]][1], delim='\\t') %>% \n",
    "    rename_with(.cols = all_of(starts_with(\"Urban_\")), .fn = ~str_replace(., \"Urban_\", \"\")) %>% \n",
    "    rename(\"Chr\" = \"chr\", \"mean\" = \"L\") %>% \n",
    "    mutate(habitat = \"Urban\",\n",
    "           is_outlier = ifelse(mean >= quantile(.$mean, prob = 0.99), 1, 0)) %>% \n",
    "    mutate(winCenter = start + ((end - start) / 2)) %>% \n",
    "    remap_chr_names()\n",
    "\n",
    "rural_lassi_df <- read_delim(snakemake@input[[\"lassip\"]][2], delim='\\t') %>% \n",
    "    rename_with(.cols = all_of(starts_with(\"Rural_\")), .fn = ~str_replace(., \"Rural_\", \"\")) %>% \n",
    "    rename(\"Chr\" = \"chr\", \"mean\" = \"L\") %>% \n",
    "    mutate(habitat = \"Rural\",\n",
    "           is_outlier = ifelse(mean >= quantile(.$mean, prob = 0.99), 1, 0)) %>% \n",
    "    mutate(winCenter = start + ((end - start) / 2)) %>% \n",
    "    remap_chr_names()\n",
    "\n",
    "write_delim(bind_rows(urban_lassi_df, rural_lassi_df), snakemake@output[[\"salti_df\"]], delim='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "01cd8da7-28fb-4884-878c-e207cd327a3a",
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_single_population_hapstat_win_manhat <- function(df, hab, stat){\n",
    "\n",
    "    df <- df %>% filter(habitat == hab)\n",
    "    \n",
    "    # Setting up cummulative genome-wide x-axis\n",
    "    data_cum <- df %>% \n",
    "        group_by(Chr) %>% \n",
    "        summarise(max_winCenter = max(winCenter)) %>% \n",
    "        mutate(winCenter_add = lag(cumsum(max_winCenter), default = 0)) %>% \n",
    "        dplyr::select(Chr, winCenter_add)\n",
    "    \n",
    "    df_mod <- df %>%\n",
    "        inner_join(data_cum, by = \"Chr\") %>% \n",
    "        mutate(winCenter_cum = winCenter + winCenter_add) \n",
    "    \n",
    "    axis_set <- df_mod %>% \n",
    "      group_by(Chr) %>% \n",
    "      summarize(center = mean(winCenter_cum))\n",
    "\n",
    "    # Get Fst outliers and genome-wide critical value\n",
    "    outliers <- df_mod %>% filter(is_outlier == 1) \n",
    "    quant <- quantile(abs(df %>% pull(mean)), probs = c(0.99))\n",
    "\n",
    "    if(hab == 'Urban'){\n",
    "        col = '#003876'\n",
    "    } else {\n",
    "        col = '#007243'\n",
    "    }\n",
    "\n",
    "    if(stat == 'nSL'){\n",
    "        ylab = '| nSL |'\n",
    "        coord = c(0, 4)\n",
    "        breaks = seq(0, 4, 1)\n",
    "    } else if (stat == 'iHS') {\n",
    "        ylab = '| iHS |'\n",
    "        coord = c(0, 3.25)\n",
    "        breaks = seq(0, 3, 1)\n",
    "    } else if (stat == \"iHH12\") {\n",
    "        ylab = 'iHH12'\n",
    "        coord = c(-0.5, 17)\n",
    "        breaks = seq(0, 16, 2)\n",
    "    } else {\n",
    "        ylab = 'Lambda'\n",
    "        coord = c(-0.5, 140)\n",
    "        breaks = seq(0, 140, 20)\n",
    "    }\n",
    "\n",
    "    # Generate Manhattan plot\n",
    "    manhat_plot <- df_mod %>%\n",
    "        filter(is_outlier != 1) %>% \n",
    "        mutate(chrom_cat = case_when(Chr == 1 ~ 'One',\n",
    "                                     Chr == 2 ~ 'Two',\n",
    "                                     Chr == 3 ~ 'One',\n",
    "                                     Chr == 4 ~ 'Two',\n",
    "                                     Chr == 5 ~ 'One',\n",
    "                                     Chr == 6 ~ 'Two',\n",
    "                                     Chr == 7 ~ 'One',\n",
    "                                     Chr == 8 ~ 'Two',\n",
    "                                     Chr == 9 ~ 'One',\n",
    "                                     Chr == 10 ~ 'Two',\n",
    "                                     Chr == 11 ~ 'One',\n",
    "                                     Chr == 12 ~ 'Two',\n",
    "                                     Chr == 13 ~ 'One',\n",
    "                                     Chr == 14 ~ 'Two',\n",
    "                                     Chr == 15 ~ 'One',\n",
    "                                     Chr == 16 ~ 'Two')) %>% \n",
    "            ggplot(aes(x = winCenter_cum, y = mean)) +\n",
    "            geom_point(shape = 21, alpha = 0.4, size = 3, aes(fill = chrom_cat, color = chrom_cat)) +\n",
    "            geom_point(data = outliers, shape = 21, alpha = 1, size = 3, color = col, fill = col) +\n",
    "            geom_hline(yintercept = quant, color = \"grey40\", linetype = \"dashed\") +\n",
    "            scale_x_continuous(label = axis_set$Chr, breaks=axis_set$center) +\n",
    "            scale_y_continuous(expand = c(0,0), breaks = breaks) +\n",
    "            coord_cartesian(ylim = coord) +\n",
    "            scale_fill_manual(values = c(\"black\", \"grey40\")) + \n",
    "            scale_color_manual(values = c(\"black\", \"grey40\")) + \n",
    "            ylab(ylab) + xlab('') +\n",
    "            ggtitle(sprintf(\"%s Windowed %s\", hab, stat)) +\n",
    "            theme_classic() +\n",
    "            theme(\n",
    "                legend.position = \"none\",\n",
    "                panel.border = element_blank(),\n",
    "                panel.grid.major.x = element_blank(),\n",
    "                panel.grid.minor.x = element_blank(),\n",
    "                axis.text = element_text(size=16),\n",
    "                axis.title = element_text(size=20),\n",
    "                plot.title = element_text(size = 20, face = \"bold\")\n",
    "              )\n",
    "    return(manhat_plot)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "48d71c8f-990f-4bc9-a700-c971ba9ef176",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 20, repr.plot.height = 6)\n",
    "urban_salti_manhat <- plot_single_population_hapstat_win_manhat(urban_lassi_df, \"Urban\", \"saltiLassi\")\n",
    "ggsave(filename = snakemake@output[[\"urban_salti_manhat\"]], plot = urban_salti_manhat, \n",
    "       height = 8, width = 20, device = \"pdf\", dpi = 600, units = \"in\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ae5d38f3-2449-48b2-ba07-ff16542e8189",
   "metadata": {},
   "outputs": [],
   "source": [
    "rural_salti_manhat <- plot_single_population_hapstat_win_manhat(rural_lassi_df, \"Rural\", \"saltiLassi\")\n",
    "ggsave(filename = snakemake@output[[\"rural_salti_manhat\"]], plot = rural_salti_manhat, \n",
    "       height = 8, width = 20, device = \"pdf\", dpi = 600, units = \"in\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "89c770a9-7ca7-4eb9-b031-b56dc92f9c8f",
   "metadata": {},
   "outputs": [],
   "source": [
    "cols <- c(\"Chr\", \"start\", \"end\", \"winCenter\", \"is_outlier\", \"m\", \"A\", \"mean\", \"habitat\")\n",
    "urban_lassi_only <- bt.intersect(urban_lassi_df %>% dplyr::select(Chr, start, end, winCenter, is_outlier, m, A, mean, habitat) %>% \n",
    "                                     filter(is_outlier == 1), \n",
    "                                 rural_lassi_df %>% dplyr::select(Chr, start, end, winCenter, is_outlier, m, A, mean, habitat) %>% \n",
    "                                     filter(is_outlier == 0), wa=T) %>% distinct()\n",
    "names(urban_lassi_only) <- cols\n",
    "rural_lassi_only <- bt.intersect(rural_lassi_df %>% dplyr::select(Chr, start, end, winCenter, is_outlier, m, A, mean, habitat) %>% \n",
    "                                     filter(is_outlier == 1), \n",
    "                                 urban_lassi_df %>% dplyr::select(Chr, start, end, winCenter, is_outlier, m, A, mean, habitat) %>% \n",
    "                                     filter(is_outlier == 0), wa=T) %>% distinct()\n",
    "names(rural_lassi_only) <- cols"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ebb4983b-ba8a-4323-ab8d-2aaeb1857414",
   "metadata": {},
   "outputs": [],
   "source": [
    "urban_lassi_xpnsl <- bt.intersect(urban_lassi_only %>% \n",
    "                                      dplyr::select(Chr, start, end, winCenter, is_outlier, m, A, mean, habitat), \n",
    "                                  obs_xpnsl_ur_df_filt %>% filter(direction == \"Urban\"), wa=T) %>% distinct()\n",
    "names(urban_lassi_xpnsl) <- cols\n",
    "rural_lassi_xpnsl <- bt.intersect(rural_lassi_only %>% \n",
    "                                      dplyr::select(Chr, start, end, winCenter, is_outlier, m, A, mean, habitat), \n",
    "                                  obs_xpnsl_ur_df_filt %>% filter(direction == \"Rural\"), wa=T) %>% distinct()\n",
    "names(rural_lassi_xpnsl) <- cols\n",
    "lassi_xpnsl <- bind_rows(urban_lassi_xpnsl, rural_lassi_xpnsl) %>% \n",
    "    distinct()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "990a537a-8baf-4524-a8eb-1407560ef17b",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 8, repr.plot.height = 6)\n",
    "salti_m_hist <- lassi_xpnsl %>% \n",
    "    group_by(habitat, m) %>% \n",
    "    summarize(count = n()) %>% \n",
    "    ungroup() %>% \n",
    "    complete(habitat, m, fill = list(count = 0)) %>% \n",
    "    ggplot(aes(x = m, y = count)) +\n",
    "        geom_bar(stat = \"identity\", position = \"dodge\", aes(color = habitat, fill = habitat)) +\n",
    "        scale_color_manual(values = c(\"#007243\", \"#003876\")) +\n",
    "        scale_fill_manual(values = c(\"#007243\", \"#003876\")) +\n",
    "        ylab(\"Number of windows\") + xlab(\"Number of sweeping haplotypes\") +\n",
    "        scale_x_continuous(breaks = seq(1, 8, 1)) +\n",
    "        theme_classic() +\n",
    "        theme(axis.text = element_text(size = 15),\n",
    "              axis.title = element_text(size = 17),\n",
    "              legend.text = element_text(size = 13),\n",
    "              legend.title = element_text(size = 15)) +\n",
    "        transp_theme\n",
    "salti_m_hist\n",
    "ggsave(filename = snakemake@output[[\"salti_m_hist\"]], plot = salti_m_hist, \n",
    "       height = 8, width = 8, device = \"pdf\", dpi = 600, units = \"in\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "41f2bca0-0f3c-4ae5-9d12-d3d362674302",
   "metadata": {},
   "outputs": [],
   "source": [
    "salti_sr_hist <- lassi_xpnsl %>% \n",
    "    mutate(sweep_type = ifelse(m > 1, \"Soft (m > 1)\", \"Hard (m = 1)\")) %>% \n",
    "    group_by(habitat, sweep_type) %>% \n",
    "    summarize(count = n()) %>% \n",
    "    ungroup() %>% \n",
    "    ggplot(aes(x = sweep_type, y = count)) +\n",
    "        geom_bar(stat = \"identity\", position = \"dodge\", aes(color = habitat, fill = habitat)) +\n",
    "        scale_color_manual(values = c(\"#007243\", \"#003876\")) +\n",
    "        scale_fill_manual(values = c(\"#007243\", \"#003876\")) +\n",
    "        ylab(\"Number of windows\") + xlab(\"Sweep architecture\") +\n",
    "        theme_classic() +\n",
    "        theme(axis.text = element_text(size = 15),\n",
    "              axis.title = element_text(size = 17),\n",
    "              legend.text = element_text(size = 13),\n",
    "              legend.title = element_text(size = 15)) +\n",
    "        transp_theme\n",
    "salti_sr_hist\n",
    "ggsave(filename = snakemake@output[[\"salti_sr_hist\"]], plot = salti_sr_hist, \n",
    "       height = 8, width = 8, device = \"pdf\", dpi = 600, units = \"in\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "43a4da70-ea04-4a9a-bc60-38fb2fef7f99",
   "metadata": {},
   "outputs": [],
   "source": [
    "chisq_table <- lassi_xpnsl %>% \n",
    "    mutate(sweep_type = ifelse(m > 1, \"Soft\", \"Hard\")) %>% \n",
    "    dplyr::select(habitat, sweep_type) %>% \n",
    "    group_by(habitat, sweep_type) %>% \n",
    "    summarize(count = n()) %>% \n",
    "    ungroup() %>% \n",
    "    pivot_wider(names_from = \"sweep_type\", values_from = \"count\") %>% \n",
    "    column_to_rownames(., var = \"habitat\") %>% \n",
    "    as.matrix()\n",
    "chisq_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "73fba0f4-cfa2-4754-b1ee-1dd3a243edd3",
   "metadata": {},
   "outputs": [],
   "source": [
    "chisq_sweep_arch <- chisq.test(chisq_table)\n",
    "chisq_sweep_arch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ee9d5b7d-5ab0-4e82-b7fd-a2292946eaec",
   "metadata": {},
   "outputs": [],
   "source": [
    "urban_A <- mean(log10(lassi_xpnsl %>% filter(habitat == \"Urban\") %>% pull(A)))\n",
    "rural_A <- mean(log10(lassi_xpnsl %>% filter(habitat == \"Rural\") %>% pull(A)))\n",
    "A_diff_obs <- urban_A - rural_A\n",
    "\n",
    "# Set seed for reproducible results\n",
    "set.seed(49)\n",
    "\n",
    "# Initialize a list to store the simulated test-statistics\n",
    "simulated_diffs <- list()\n",
    "nreps = 1000\n",
    "\n",
    "for(i in 1:nreps){\n",
    "\n",
    "    # Create temporary dataframe to permute so we don't modify the original\n",
    "    reshuffled <- lassi_xpnsl \n",
    "    \n",
    "    # Permute the width column with the 'sample()' function. \n",
    "    reshuffled$A <- sample(reshuffled$A, size = nrow(reshuffled), replace = FALSE)\n",
    "    \n",
    "    # Calculate the means for each habitat\n",
    "    mean_urban_sim <- mean(log10(reshuffled %>% filter(habitat == 'Urban') %>% pull(A)))\n",
    "    mean_rural_sim <- mean(log10(reshuffled %>% filter(habitat == 'Rural') %>% pull(A)))\n",
    "    \n",
    "    # Calculate to difference between simulated urban and rural sweep widths\n",
    "    A_diff_sim <- mean_urban_sim - mean_rural_sim\n",
    "\n",
    "    # Append simulated mean difference to list\n",
    "    simulated_diffs[i] <- A_diff_sim\n",
    "}    \n",
    "\n",
    "# Unlist simulated means list into numeric vector\n",
    "simulated_diffs <- unlist(simulated_diffs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "072c5d1b-19cc-4cf8-b207-1e60408513b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "logA_perm_plot <- ggplot() +\n",
    "    geom_histogram(aes(x = simulated_diffs), bins = 30, \n",
    "                   fill = \"grey\", alpha = 0.4, colour = \"black\") +\n",
    "    geom_vline(xintercept = A_diff_obs, linewidth = 1, \n",
    "               linetype = \"dashed\", colour = \"black\") + \n",
    "    ylab(\"Number of iterations\") + \n",
    "    xlab(\"Simulated urban-rural difference in log(A) between habitats\") +\n",
    "    theme_classic() +\n",
    "    theme(axis.title = element_text(size = 17),\n",
    "          axis.text = element_text(size = 15)) +\n",
    "    transp_theme\n",
    "logA_perm_plot\n",
    "\n",
    "ggsave(filename = snakemake@output[[\"logA_perm_plot\"]], plot = logA_perm_plot, \n",
    "       height = 8, width = 8, device = \"pdf\", dpi = 600, units = \"in\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2adb0737-651a-4787-90d9-113a48092a38",
   "metadata": {},
   "outputs": [],
   "source": [
    "logA_ur_density <- lassi_xpnsl %>% \n",
    "    group_by(habitat) %>% \n",
    "    ggplot(aes(x = log10(A))) +\n",
    "        geom_density(aes(color = habitat, fill = habitat), alpha = 0.7) +\n",
    "        scale_color_manual(values = c(\"#007243\", \"#003876\")) +\n",
    "        scale_fill_manual(values = c(\"#007243\", \"#003876\")) +\n",
    "        ylab(\"Density\") + \n",
    "        xlab(\"Width of selective sweep [log(A)]\") +\n",
    "        theme_classic() +\n",
    "        theme(axis.title = element_text(size = 17),\n",
    "              axis.text = element_text(size = 15)) +\n",
    "        transp_theme\n",
    "logA_ur_density\n",
    "\n",
    "ggsave(filename = snakemake@output[[\"logA_ur_density\"]], plot = logA_ur_density, \n",
    "       height = 8, width = 8, device = \"pdf\", dpi = 600, units = \"in\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "90a7e6ac-ff77-47cb-a989-0c4b1b862203",
   "metadata": {},
   "source": [
    "## nSL\n",
    "\n",
    "- I estimated nSL in the same 50 Kb windows as above, separately in urban and rural habitats\n",
    "- I'll do the same thing as I did above, minus the permutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f401444c-5165-4567-9ac4-6b53ff8985a0",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "nsl_df <- read_delim(snakemake@input[[\"win_nsl\"]], delim='\\t') %>% \n",
    "    remap_chr_names()\n",
    "head(nsl_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6af91a81-22a2-4381-95a7-01a5b8cf0312",
   "metadata": {},
   "outputs": [],
   "source": [
    "nsl_df %>% \n",
    "    ggplot(aes(x = n)) +\n",
    "        geom_histogram(bins = 50, color = \"black\", fill = \"white\") +\n",
    "        ylab(\"Number of windows\") + xlab(\"Number of sites\") +\n",
    "        geom_vline(xintercept = 25, color = \"red\", linetype = \"dashed\", linewidth = 1) +\n",
    "        theme_classic() +\n",
    "        theme(axis.text = element_text(size = 15),\n",
    "              axis.title = element_text(size = 17))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bdcff602-8c44-4d40-9dd6-f244c53216bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to assign outliers to nSL windows\n",
    "assign_nsl_outliers <- function(df, thresh = nsl_nsites_thresh){\n",
    "\n",
    "    df_filt <- df %>%\n",
    "        filter(n >= thresh)\n",
    "\n",
    "    # Get critical values for mean XP-nSL score and proportions greater or lesser than 2 and -2, respectively\n",
    "    score_quant <- quantile(df_filt %>% pull(mean), probs = 0.99)\n",
    "    gtprop_quant <- quantile(df_filt %>% pull(gt_frac), probs = 0.99)\n",
    "\n",
    "    # Identify outliers and add as categorical variable to windows dataframe\n",
    "    df_filt <- df_filt %>%\n",
    "        mutate(score_outlier = ifelse(mean >= score_quant, 1, 0),\n",
    "               gtprop_outlier = ifelse(gt_frac >= gtprop_quant, 1, 0),\n",
    "               is_outlier = ifelse(score_outlier == 1 & gtprop_outlier == 1, 1, 0))\n",
    "\n",
    "    return(df_filt)\n",
    "}\n",
    "\n",
    "nsl_nsites_thresh <- 25\n",
    "nsl_df_filt <- nsl_df %>% \n",
    "    group_split(habitat) %>% \n",
    "    purrr::map_dfr(., assign_nsl_outliers, thresh = nsl_nsites_thresh)\n",
    "\n",
    "sprintf(\"%s of %s nSL windows remaining after removing those with less than %s sites\", nrow(nsl_df_filt), nrow(nsl_df), nsl_nsites_thresh)\n",
    "\n",
    "head(nsl_df_filt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "73a12718-346a-461c-9f4c-88f9b6783d16",
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_single_population_hapstat_win_manhat <- function(df, hab, stat){\n",
    "\n",
    "    df <- df %>% filter(habitat == hab)\n",
    "    \n",
    "    # Setting up cummulative genome-wide x-axis\n",
    "    data_cum <- df %>% \n",
    "        group_by(Chr) %>% \n",
    "        summarise(max_winCenter = max(winCenter)) %>% \n",
    "        mutate(winCenter_add = lag(cumsum(max_winCenter), default = 0)) %>% \n",
    "        dplyr::select(Chr, winCenter_add)\n",
    "    \n",
    "    df_mod <- df %>%\n",
    "        inner_join(data_cum, by = \"Chr\") %>% \n",
    "        mutate(winCenter_cum = winCenter + winCenter_add) \n",
    "    \n",
    "    axis_set <- df_mod %>% \n",
    "        group_by(Chr) %>% \n",
    "        summarize(center = mean(winCenter_cum))\n",
    "\n",
    "    # Get Fst outliers and genome-wide critical value\n",
    "    outliers <- df_mod %>% filter(is_outlier == 1) \n",
    "    quant <- quantile(abs(df %>% pull(mean)), probs = c(0.99))\n",
    "\n",
    "    if(hab == 'Urban'){\n",
    "        col = '#003876'\n",
    "    } else {\n",
    "        col = '#007243'\n",
    "    }\n",
    "\n",
    "    if(stat == 'nSL'){\n",
    "        ylab = '| nSL |'\n",
    "        coord = c(0, 4)\n",
    "        breaks = seq(0, 4, 1)\n",
    "    } else if (stat == 'iHS') {\n",
    "        ylab = '| iHS |'\n",
    "        coord = c(0, 3.25)\n",
    "        breaks = seq(0, 3, 1)\n",
    "    } else {\n",
    "        ylab = 'iHH12'\n",
    "        coord = c(-0.5, 17)\n",
    "        breaks = seq(0, 16, 2)\n",
    "    }\n",
    "\n",
    "    # Generate Manhattan plot\n",
    "    manhat_plot <- df_mod %>%\n",
    "        filter(is_outlier != 1) %>% \n",
    "        mutate(chrom_cat = case_when(Chr == 1 ~ 'One',\n",
    "                                     Chr == 2 ~ 'Two',\n",
    "                                     Chr == 3 ~ 'One',\n",
    "                                     Chr == 4 ~ 'Two',\n",
    "                                     Chr == 5 ~ 'One',\n",
    "                                     Chr == 6 ~ 'Two',\n",
    "                                     Chr == 7 ~ 'One',\n",
    "                                     Chr == 8 ~ 'Two',\n",
    "                                     Chr == 9 ~ 'One',\n",
    "                                     Chr == 10 ~ 'Two',\n",
    "                                     Chr == 11 ~ 'One',\n",
    "                                     Chr == 12 ~ 'Two',\n",
    "                                     Chr == 13 ~ 'One',\n",
    "                                     Chr == 14 ~ 'Two',\n",
    "                                     Chr == 15 ~ 'One',\n",
    "                                     Chr == 16 ~ 'Two')) %>% \n",
    "            ggplot(aes(x = winCenter_cum, y = mean)) +\n",
    "            geom_point(shape = 21, alpha = 0.4, size = 3, aes(fill = chrom_cat, color = chrom_cat)) +\n",
    "            geom_point(data = outliers, shape = 21, alpha = 1, size = 3, color = col, fill = col) +\n",
    "            geom_hline(yintercept = quant, color = \"grey40\", linetype = \"dashed\") +\n",
    "            scale_x_continuous(label = axis_set$Chr, breaks=axis_set$center) +\n",
    "            scale_y_continuous(expand = c(0,0), breaks = breaks) +\n",
    "            coord_cartesian(ylim = coord) +\n",
    "            scale_fill_manual(values = c(\"black\", \"grey40\")) + \n",
    "            scale_color_manual(values = c(\"black\", \"grey40\")) + \n",
    "            ylab(ylab) + xlab('') +\n",
    "            ggtitle(sprintf(\"%s Windowed %s\", hab, stat)) +\n",
    "            theme_classic() +\n",
    "            theme(\n",
    "                legend.position = \"none\",\n",
    "                panel.border = element_blank(),\n",
    "                panel.grid.major.x = element_blank(),\n",
    "                panel.grid.minor.x = element_blank(),\n",
    "                axis.text = element_text(size=16),\n",
    "                axis.title = element_text(size=20),\n",
    "                plot.title = element_text(size = 20, face = \"bold\")\n",
    "              )\n",
    "    return(manhat_plot)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a0e27f8c-3eb6-4750-ad35-cc185f7bc339",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 20, repr.plot.height = 6)\n",
    "rural_nsl_manhat <- plot_single_population_hapstat_win_manhat(nsl_df_filt, \"Rural\", stat = 'nSL')\n",
    "rural_nsl_manhat\n",
    "ggsave(filename = snakemake@output[[\"rur_nsl_manhat\"]], plot = rural_nsl_manhat, \n",
    "       height = 8, width = 20, device = \"pdf\", dpi = 600, units = \"in\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "23e7c98a-c741-4e0c-9328-701105cb79d5",
   "metadata": {},
   "outputs": [],
   "source": [
    "urban_nsl_manhat <- plot_single_population_hapstat_win_manhat(nsl_df_filt, \"Urban\", stat = 'nSL')\n",
    "urban_nsl_manhat\n",
    "ggsave(filename = snakemake@output[[\"urb_nsl_manhat\"]], plot = urban_nsl_manhat, \n",
    "       height = 8, width = 20, device = \"pdf\", dpi = 600, units = \"in\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d123430e-830a-4bee-ad2a-f064834cdb7d",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "nsl_df_filt %>% \n",
    "    filter(is_outlier == 1) %>% \n",
    "    write_delim(snakemake@output[[\"nsl_df\"]], delim=\"\\t\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5f5841f2-9531-4bd1-ae7b-fc3697d69506",
   "metadata": {},
   "source": [
    "## iHS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "42afdd3e-ef89-43cd-bc79-3b83fe6de0cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "ihs_df <- read_delim(snakemake@input[[\"win_ihs\"]], delim='\\t') %>% \n",
    "    remap_chr_names()\n",
    "head(ihs_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ed8f08fd-d3e7-4deb-9bab-5d09ca37b684",
   "metadata": {},
   "outputs": [],
   "source": [
    "ihs_df_filt <- ihs_df %>% \n",
    "    group_split(habitat) %>% \n",
    "    purrr::map_dfr(., assign_nsl_outliers, thresh = nsl_nsites_thresh)\n",
    "\n",
    "sprintf(\"%s of %s iHS windows remaining after removing those with less than %s sites\", nrow(ihs_df_filt), nrow(ihs_df), nsl_nsites_thresh)\n",
    "\n",
    "head(ihs_df_filt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a58f3335-0501-4940-b774-a3cd5cd1fc7e",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "options(repr.plot.width = 20, repr.plot.height = 6)\n",
    "rural_ihs_manhat <- plot_single_population_hapstat_win_manhat(ihs_df_filt, \"Rural\", stat = 'iHS')\n",
    "rural_ihs_manhat\n",
    "ggsave(filename = snakemake@output[[\"rur_ihs_manhat\"]], plot = rural_ihs_manhat, \n",
    "       height = 8, width = 20, device = \"pdf\", dpi = 600, units = \"in\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "250472d8-8fb7-4550-bb0b-e45987351f0e",
   "metadata": {},
   "outputs": [],
   "source": [
    "urban_ihs_manhat <- plot_single_population_hapstat_win_manhat(ihs_df_filt, \"Urban\", stat = 'iHS')\n",
    "urban_ihs_manhat\n",
    "ggsave(filename = snakemake@output[[\"urb_ihs_manhat\"]], plot = urban_ihs_manhat, \n",
    "       height = 8, width = 20, device = \"pdf\", dpi = 600, units = \"in\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "de4b44de-5fc2-4097-a9d7-110169102f69",
   "metadata": {},
   "outputs": [],
   "source": [
    "ihs_df_filt %>% \n",
    "    filter(is_outlier == 1) %>% \n",
    "    write_delim(snakemake@output[[\"ihs_df\"]], delim=\"\\t\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "07be8c88-7bd8-4507-ad25-14b3966531d0",
   "metadata": {},
   "source": [
    "## iHH12"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4e918166-8e0c-49d2-868d-2975f4fa2f02",
   "metadata": {},
   "outputs": [],
   "source": [
    "ihh12_df <- read_delim(snakemake@input[[\"win_ihh12\"]], delim='\\t') %>% \n",
    "    remap_chr_names()\n",
    "head(ihh12_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "70e7b205-2deb-4b1e-a791-34a90e71dd9a",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "ihh12_df_filt <- ihh12_df %>% \n",
    "    group_split(habitat) %>% \n",
    "    purrr::map_dfr(., assign_nsl_outliers, thresh = nsl_nsites_thresh)\n",
    "\n",
    "sprintf(\"%s of %s iHH12 windows remaining after removing those with less than %s sites\", nrow(ihh12_df_filt), nrow(ihh12_df), nsl_nsites_thresh)\n",
    "\n",
    "head(ihh12_df_filt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8135a452-155d-4789-8f39-a795f4215f55",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 20, repr.plot.height = 6)\n",
    "urban_ihh12_manhat <- plot_single_population_hapstat_win_manhat(ihh12_df_filt, \"Urban\", stat = 'iHH12')\n",
    "urban_ihh12_manhat\n",
    "ggsave(filename = snakemake@output[[\"urb_ihh12_manhat\"]], plot = urban_ihh12_manhat, \n",
    "       height = 8, width = 20, device = \"pdf\", dpi = 600, units = \"in\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0ffa328a-689d-4fab-873c-9aa3fd95e8cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "rural_ihh12_manhat <- plot_single_population_hapstat_win_manhat(ihh12_df_filt, \"Rural\", stat = 'iHH12')\n",
    "rural_ihh12_manhat\n",
    "ggsave(filename = snakemake@output[[\"rur_ihh12_manhat\"]], plot = rural_ihh12_manhat, \n",
    "       height = 8, width = 20, device = \"pdf\", dpi = 600, units = \"in\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "37f0c710-96a4-4c01-aee6-3550f5a87205",
   "metadata": {},
   "outputs": [],
   "source": [
    "ihh12_df_filt %>% \n",
    "    filter(is_outlier == 1) %>% \n",
    "    write_delim(snakemake@output[[\"ihh12_df\"]], delim=\"\\t\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "837cda8f-9385-4220-bf43-d9a30cf62cc1",
   "metadata": {},
   "source": [
    "## Fst"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "22880300-fc39-4131-903d-261804fb5f29",
   "metadata": {},
   "outputs": [],
   "source": [
    "gt_fst_df <- read_delim(snakemake@input[[\"gt_win_fst\"]]) %>% \n",
    "    filter(pop1 == \"Urban\" & pop2 == \"Rural\") %>%\n",
    "    mutate(winCenter = (window_pos_1 - 1) + 25000) %>% \n",
    "    rename(\"Chr\" = \"chromosome\",\n",
    "           \"start\" = \"window_pos_1\",\n",
    "           \"end\" = \"window_pos_2\") %>% \n",
    "    remap_chr_names()\n",
    "head(gt_fst_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "47dbc572-e034-4ecc-bc72-ff73ed7e2f78",
   "metadata": {},
   "outputs": [],
   "source": [
    "gt_fst_df_filt <- gt_fst_df %>% \n",
    "    filter(no_snps >= xpnsl_nsites_thresh) %>% \n",
    "    mutate(avg_hudson_fst = ifelse(avg_hudson_fst < 0, 0, avg_hudson_fst))\n",
    "\n",
    "gt_fst_quant <- quantile(gt_fst_df_filt %>% pull(avg_hudson_fst), probs = 0.99)\n",
    "\n",
    "gt_fst_df_filt <- gt_fst_df_filt %>% \n",
    "    mutate(is_outlier = ifelse(avg_hudson_fst >= gt_fst_quant, 1, 0))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "df147c43-5881-425c-8e06-81ce290b7750",
   "metadata": {},
   "outputs": [],
   "source": [
    "range(gt_fst_df_filt$avg_hudson_fst)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7b168c46-edde-4f3f-a5b7-292c784c067d",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 20, repr.plot.height = 4)\n",
    "\n",
    "# Setting up cummulative genome-wide x-axis\n",
    "data_cum <- gt_fst_df_filt %>% \n",
    "    group_by(Chr) %>% \n",
    "    summarise(max_winCenter = max(winCenter)) %>% \n",
    "    mutate(winCenter_add = lag(cumsum(max_winCenter), default = 0)) %>% \n",
    "    dplyr::select(Chr, winCenter_add)\n",
    "\n",
    "gt_fst_df_filt_mod <- gt_fst_df_filt %>%\n",
    "    inner_join(data_cum, by = \"Chr\") %>% \n",
    "    mutate(winCenter_cum = winCenter + winCenter_add) \n",
    "\n",
    "axis_set <- gt_fst_df_filt_mod %>% \n",
    "    group_by(Chr) %>% \n",
    "    summarize(center = mean(winCenter_cum))\n",
    "\n",
    "# Get Fst outliers for plotting as separate layers\n",
    "gt_fst_not_outlier <- gt_fst_df_filt_mod %>% filter(is_outlier == 0)\n",
    "gt_fst_outliers <- gt_fst_df_filt_mod %>% filter(is_outlier == 1)\n",
    "\n",
    "gt_fst_not_outlier <- gt_fst_not_outlier %>%\n",
    "        mutate(chrom_cat = case_when(Chr == 1 ~ 'One',\n",
    "                                     Chr == 2 ~ 'Two',\n",
    "                                     Chr == 3 ~ 'One',\n",
    "                                     Chr == 4 ~ 'Two',\n",
    "                                     Chr == 5 ~ 'One',\n",
    "                                     Chr == 6 ~ 'Two',\n",
    "                                     Chr == 7 ~ 'One',\n",
    "                                     Chr == 8 ~ 'Two',\n",
    "                                     Chr == 9 ~ 'One',\n",
    "                                     Chr == 10 ~ 'Two',\n",
    "                                     Chr == 11 ~ 'One',\n",
    "                                     Chr == 12 ~ 'Two',\n",
    "                                     Chr == 13 ~ 'One',\n",
    "                                     Chr == 14 ~ 'Two',\n",
    "                                     Chr == 15 ~ 'One',\n",
    "                                     Chr == 16 ~ 'Two'))\n",
    "\n",
    "gt_fst_manhat <- ggplot() +\n",
    "        geom_point(data = gt_fst_not_outlier, shape = 21, alpha = 0.4, size = 3, \n",
    "                   aes(x = winCenter_cum, y = avg_hudson_fst, fill = chrom_cat, color = chrom_cat)) +\n",
    "        geom_point(data = gt_fst_outliers, shape = 21, alpha = 1, size = 3, color = 'red', fill = 'red',\n",
    "                   aes(x = winCenter_cum, y = avg_hudson_fst)) +\n",
    "        geom_hline(yintercept = gt_fst_quant, color = \"grey40\", linetype = \"dashed\") +\n",
    "        scale_x_continuous(label = axis_set$Chr, breaks = axis_set$center) +\n",
    "        scale_y_continuous(expand = c(0,0), breaks = seq(0, 0.06, 0.02)) +\n",
    "        coord_cartesian(ylim = c(0, 0.06)) +\n",
    "        scale_fill_manual(values = c(\"black\", \"grey40\")) + \n",
    "        scale_color_manual(values = c(\"black\", \"grey40\")) + \n",
    "        ylab(\"Urban-Rural Hudson's Fst\") + xlab('Chromosomes') +\n",
    "        theme_classic() +\n",
    "        theme(\n",
    "            legend.position = \"none\",\n",
    "            panel.border = element_blank(),\n",
    "            panel.grid.major.x = element_blank(),\n",
    "            panel.grid.minor.x = element_blank(),\n",
    "            axis.text = element_text(size=16),\n",
    "            axis.title = element_text(size=20)) +\n",
    "        transp_theme\n",
    "\n",
    "options(repr.plot.width = 20, repr.plot.height = 6)\n",
    "gt_fst_manhat\n",
    "ggsave(filename = snakemake@output[[\"gt_fst_manhat\"]], plot = gt_fst_manhat, \n",
    "       height = 4, width = 20, device = \"pdf\", dpi = 600, units = \"in\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "704e9867-1fa1-40b2-9d7b-139a65f4ba5c",
   "metadata": {},
   "outputs": [],
   "source": [
    "gt_fst_df_filt %>% \n",
    "    filter(is_outlier == 1) %>% \n",
    "    write_delim(snakemake@output[[\"gt_fst_df\"]], delim=\"\\t\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1481b363-a8ce-4037-ae28-50a221460450",
   "metadata": {},
   "source": [
    "## Tajima's D"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5c655648-feb4-4c3b-8343-42cf4ed3bad7",
   "metadata": {},
   "outputs": [],
   "source": [
    "win_sfs_stats <- read_delim(snakemake@input[[\"win_sfs_fst\"]], delim = \"\\t\") %>% \n",
    "    remap_chr_names() %>%\n",
    "    rename(\"winCenter\" = \"WinCenter\") %>% \n",
    "    dplyr::select(Chr, start, end, winCenter, nSites_Urban, nSites_Rural, Tajima_Urban, Tajima_Rural, delta_td_ur) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e5500976-fd68-4a19-8ac2-2eeb150431a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "td_diff_quant <- quantile(win_sfs_stats$delta_td_ur, probs = c(0.01, 0.99))\n",
    "win_sfs_stats_filt <- win_sfs_stats %>% \n",
    "    mutate(is_outlier = case_when(delta_td_ur <= td_diff_quant[1] ~ 1,\n",
    "                                 delta_td_ur >= td_diff_quant[2] ~ 1,\n",
    "                                 TRUE ~ 0))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "51df213c-7118-4183-b44a-59c17ace8f59",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 20, repr.plot.height = 4)\n",
    "\n",
    "# Setting up cummulative genome-wide x-axis\n",
    "data_cum <- win_sfs_stats_filt %>% \n",
    "    group_by(Chr) %>% \n",
    "    summarise(max_winCenter = max(winCenter)) %>% \n",
    "    mutate(winCenter_add = lag(cumsum(max_winCenter), default = 0)) %>% \n",
    "    dplyr::select(Chr, winCenter_add)\n",
    "\n",
    "win_sfs_stats_mod <- win_sfs_stats_filt %>%\n",
    "    inner_join(data_cum, by = \"Chr\") %>% \n",
    "    mutate(winCenter_cum = winCenter + winCenter_add) \n",
    "\n",
    "axis_set <- win_sfs_stats_mod %>% \n",
    "    group_by(Chr) %>% \n",
    "    summarize(center = mean(winCenter_cum))\n",
    "\n",
    "td_not_outlier <- win_sfs_stats_mod %>% filter(is_outlier == 0)\n",
    "td_outliers <- win_sfs_stats_mod %>% filter(is_outlier == 1)\n",
    "\n",
    "td_not_outlier <- td_not_outlier %>%\n",
    "        mutate(chrom_cat = case_when(Chr == 1 ~ 'One',\n",
    "                                     Chr == 2 ~ 'Two',\n",
    "                                     Chr == 3 ~ 'One',\n",
    "                                     Chr == 4 ~ 'Two',\n",
    "                                     Chr == 5 ~ 'One',\n",
    "                                     Chr == 6 ~ 'Two',\n",
    "                                     Chr == 7 ~ 'One',\n",
    "                                     Chr == 8 ~ 'Two',\n",
    "                                     Chr == 9 ~ 'One',\n",
    "                                     Chr == 10 ~ 'Two',\n",
    "                                     Chr == 11 ~ 'One',\n",
    "                                     Chr == 12 ~ 'Two',\n",
    "                                     Chr == 13 ~ 'One',\n",
    "                                     Chr == 14 ~ 'Two',\n",
    "                                     Chr == 15 ~ 'One',\n",
    "                                     Chr == 16 ~ 'Two'))\n",
    "\n",
    "tajima_manhat <- win_sfs_stats_mod %>% \n",
    "    ggplot(aes(x = winCenter_cum, y = delta_td_ur)) +\n",
    "        geom_point(data = td_not_outlier, shape = 21, alpha = 0.4, size = 3, \n",
    "                   aes(x = winCenter_cum, y = delta_td_ur, fill = chrom_cat, color = chrom_cat)) +\n",
    "        geom_point(data = td_outliers, shape = 21, alpha = 1, size = 3, color = 'red', fill = 'red',\n",
    "                   aes(x = winCenter_cum, y = delta_td_ur)) +\n",
    "        # geom_smooth(aes(group = Chr), method = \"loess\", span = 0.1, se = FALSE, color = \"yellow\",\n",
    "        #             linewidth = 0.75) +\n",
    "        scale_x_continuous(label = axis_set$Chr, breaks = axis_set$center) +\n",
    "        scale_color_manual(values = c(\"black\", \"grey40\")) + \n",
    "        scale_fill_manual(values = c(\"black\", \"grey40\")) + \n",
    "        ylab(bquote(Tajima[Urban]~-~Tajima[Rural])) + xlab('Chromosomes') +\n",
    "        theme_classic() +\n",
    "        theme(\n",
    "            legend.position = \"none\",\n",
    "            panel.border = element_blank(),\n",
    "            panel.grid.major.x = element_blank(),\n",
    "            panel.grid.minor.x = element_blank(),\n",
    "            axis.text = element_text(size=16),\n",
    "            axis.title = element_text(size=20)) +\n",
    "        transp_theme\n",
    "tajima_manhat\n",
    "\n",
    "ggsave(filename = snakemake@output[[\"tajima_manhat\"]], plot = tajima_manhat, \n",
    "       height = 4, width = 20, device = \"pdf\", dpi = 600, units = \"in\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6ae67849-4b4b-4cd4-9742-4a677d6c1b6c",
   "metadata": {},
   "source": [
    "## LD"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "61ef3411-6588-4454-b17c-cb831b00e016",
   "metadata": {},
   "outputs": [],
   "source": [
    "win_ld_df <- read_delim(snakemake@input[[\"win_ld\"]], delim = \"\\t\") %>% \n",
    "    rename(\"ld\" = \"mean\") %>% \n",
    "    remap_chr_names()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "82bb9ccb-bc20-4800-b922-55ce5ec536fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 20, repr.plot.height = 4)\n",
    "\n",
    "# Setting up cummulative genome-wide x-axis\n",
    "data_cum <- win_ld_df %>% \n",
    "    group_by(Chr) %>% \n",
    "    summarise(max_winCenter = max(winCenter)) %>% \n",
    "    mutate(winCenter_add = lag(cumsum(max_winCenter), default = 0)) %>% \n",
    "    dplyr::select(Chr, winCenter_add)\n",
    "\n",
    "win_ld_df_mod <- win_ld_df %>%\n",
    "    inner_join(data_cum, by = \"Chr\") %>% \n",
    "    mutate(winCenter_cum = winCenter + winCenter_add) \n",
    "\n",
    "axis_set <- win_ld_df_mod %>% \n",
    "    group_by(Chr) %>% \n",
    "    summarize(center = mean(winCenter_cum))\n",
    "\n",
    "win_ld_df_mod <- win_ld_df_mod %>%\n",
    "        mutate(chrom_cat = case_when(Chr == 1 ~ 'One',\n",
    "                                     Chr == 2 ~ 'Two',\n",
    "                                     Chr == 3 ~ 'One',\n",
    "                                     Chr == 4 ~ 'Two',\n",
    "                                     Chr == 5 ~ 'One',\n",
    "                                     Chr == 6 ~ 'Two',\n",
    "                                     Chr == 7 ~ 'One',\n",
    "                                     Chr == 8 ~ 'Two',\n",
    "                                     Chr == 9 ~ 'One',\n",
    "                                     Chr == 10 ~ 'Two',\n",
    "                                     Chr == 11 ~ 'One',\n",
    "                                     Chr == 12 ~ 'Two',\n",
    "                                     Chr == 13 ~ 'One',\n",
    "                                     Chr == 14 ~ 'Two',\n",
    "                                     Chr == 15 ~ 'One',\n",
    "                                     Chr == 16 ~ 'Two'))\n",
    "\n",
    "ld_manhat <- win_ld_df_mod %>% \n",
    "    ggplot(aes(x = winCenter_cum, y = ld,)) +\n",
    "        scale_x_continuous(label = axis_set$Chr, breaks = axis_set$center) +\n",
    "        geom_smooth(data = . %>% filter(habitat == \"Urban\"), aes(group = Chr), \n",
    "                    method = \"loess\", span = 0.1, se = FALSE, color = \"#007243\",\n",
    "                    linewidth = 0.75) +\n",
    "        geom_smooth(data = . %>% filter(habitat == \"Rural\"), aes(group = Chr), \n",
    "                    method = \"loess\", span = 0.1, se = FALSE, color = \"#003876\",\n",
    "                    linewidth = 0.75) +\n",
    "        ylab(bquote(LD~(r^2))) + xlab('Chromosomes') +\n",
    "        theme_classic() +\n",
    "        theme(\n",
    "            legend.position = \"none\",\n",
    "            panel.border = element_blank(),\n",
    "            panel.grid.major.x = element_blank(),\n",
    "            panel.grid.minor.x = element_blank(),\n",
    "            axis.text = element_text(size=16),\n",
    "            axis.title = element_text(size=20)) +\n",
    "        transp_theme\n",
    "ld_manhat\n",
    "ggsave(filename = snakemake@output[[\"ld_manhat\"]], plot = ld_manhat, \n",
    "       height = 4, width = 20, device = \"pdf\", dpi = 600, units = \"in\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "07d13bb0-d741-4034-909b-8deb14faf829",
   "metadata": {},
   "source": [
    "## Recombination rate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1ce8ad71-2e9a-43d5-891d-8584da19f787",
   "metadata": {},
   "outputs": [],
   "source": [
    "gen_map <- read_delim(snakemake@input[[\"gen_map\"]], delim=\"\\t\") %>% \n",
    "    rename(\"Chr\" = \"chrom\") %>% \n",
    "    remap_chr_names()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "24e33932-fc02-40ee-806f-e16489811ea0",
   "metadata": {},
   "outputs": [],
   "source": [
    "gen_map_with_rates <- gen_map %>% \n",
    "    group_by(Chr) %>% \n",
    "    mutate(pos_diff_Mbp = (pos - lag(pos, default = first(pos))) / 1e6,\n",
    "           cM_diff = cM - lag(cM, default = first(cM)),\n",
    "           rate_cM_Mbp = ifelse(pos_diff_Mbp > 0, cM_diff / pos_diff_Mbp, NA))\n",
    "\n",
    "head(gen_map_with_rates)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "19f27e11-902f-4f24-bc3d-683e0420f7cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "calculate_windowed_recomb <- function(df, window_size, step){\n",
    "    chrom <- df %>% pull(Chr) %>% unique()\n",
    "    winStarts <- seq(from = 0, to = max(df$pos) + window_size, by = step)\n",
    "    mat <- matrix(0, nrow = length(winStarts), ncol = 6)\n",
    "    for(i in 1:length(winStarts)){\n",
    "        start <- winStarts[i]\n",
    "        end <- start + step\n",
    "        df_filt <- df %>% filter(pos >= start & pos < end)\n",
    "        winID <- i\n",
    "        winCenter <- start + (step / 2)\n",
    "        mean <- suppressWarnings(mean(df_filt$rate_cM_Mbp, na.rm = TRUE))\n",
    "        stats <- c(chrom, winID, start, end, winCenter, mean)\n",
    "        mat[i, ] <- stats\n",
    "    }\n",
    "    stats_df <- as.data.frame(mat)\n",
    "    names(stats_df) <- c(\"Chr\", \"winID\", \"start\", \"end\", \"winCenter\", \"mean\")\n",
    "    return(stats_df)\n",
    "}\n",
    "\n",
    "win_recomb <- gen_map_with_rates %>% \n",
    "    ungroup() %>% \n",
    "    group_split(Chr) %>% \n",
    "    purrr::map_dfr(., calculate_windowed_recomb, window_size = 50000, step = 50000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ab77269c-002b-40f8-a4b0-c2817b47d531",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 20, repr.plot.height = 4)\n",
    "\n",
    "# Setting up cummulative genome-wide x-axis\n",
    "data_cum <- win_recomb %>% \n",
    "    group_by(Chr) %>% \n",
    "    summarise(max_winCenter = max(winCenter)) %>% \n",
    "    mutate(winCenter_add = lag(cumsum(max_winCenter), default = 0)) %>% \n",
    "    dplyr::select(Chr, winCenter_add)\n",
    "\n",
    "win_recomb_mod <- win_recomb %>%\n",
    "    inner_join(data_cum, by = \"Chr\") %>% \n",
    "    mutate(winCenter_cum = winCenter + winCenter_add) \n",
    "\n",
    "axis_set <- win_recomb_mod %>% \n",
    "    group_by(Chr) %>% \n",
    "    summarize(center = mean(winCenter_cum))\n",
    "\n",
    "win_recomb_mod <- win_recomb_mod %>%\n",
    "        mutate(chrom_cat = case_when(Chr == 1 ~ 'One',\n",
    "                                     Chr == 2 ~ 'Two',\n",
    "                                     Chr == 3 ~ 'One',\n",
    "                                     Chr == 4 ~ 'Two',\n",
    "                                     Chr == 5 ~ 'One',\n",
    "                                     Chr == 6 ~ 'Two',\n",
    "                                     Chr == 7 ~ 'One',\n",
    "                                     Chr == 8 ~ 'Two',\n",
    "                                     Chr == 9 ~ 'One',\n",
    "                                     Chr == 10 ~ 'Two',\n",
    "                                     Chr == 11 ~ 'One',\n",
    "                                     Chr == 12 ~ 'Two',\n",
    "                                     Chr == 13 ~ 'One',\n",
    "                                     Chr == 14 ~ 'Two',\n",
    "                                     Chr == 15 ~ 'One',\n",
    "                                     Chr == 16 ~ 'Two'))\n",
    "\n",
    "recomb_manhat <- win_recomb_mod %>% \n",
    "    ggplot(aes(x = winCenter_cum, y = mean)) +\n",
    "        scale_x_continuous(label = axis_set$Chr, breaks = axis_set$center) +\n",
    "        geom_line(linewidth = 1.5, aes(color = chrom_cat)) +\n",
    "        ylab(\"Recombination rate (cM/Mb)\") + xlab('Chromosomes') +\n",
    "        scale_color_manual(values = c(\"black\", \"grey40\")) + \n",
    "        coord_cartesian(ylim = c(0, 145)) +\n",
    "        scale_y_continuous(breaks = seq(0, 140, 20)) +\n",
    "        theme_classic() +\n",
    "        theme(\n",
    "            legend.position = \"none\",\n",
    "            panel.border = element_blank(),\n",
    "            panel.grid.major.x = element_blank(),\n",
    "            panel.grid.minor.x = element_blank(),\n",
    "            axis.text = element_text(size=16),\n",
    "            axis.title = element_text(size=20)) +\n",
    "        transp_theme\n",
    "recomb_manhat\n",
    "ggsave(filename = snakemake@output[[\"recomb_manhat\"]], plot = recomb_manhat, \n",
    "       height = 4, width = 20, device = \"pdf\", dpi = 600, units = \"in\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "950a91d2-04c3-4fb6-9fbe-35bf94d5c6cd",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "xpnsl_recomb_df <- obs_xpnsl_ur_df_filt %>% \n",
    "    filter(direction != \"None\") %>% \n",
    "    dplyr::select(Chr, winCenter, mean) %>% \n",
    "    rename(\"xpnsl\" = \"mean\") %>% \n",
    "    left_join(., win_recomb, by = c(\"Chr\", \"winCenter\")) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4cce7ff2-4860-46ce-bac9-035b470c1681",
   "metadata": {},
   "outputs": [],
   "source": [
    "summary(lm(xpnsl ~ mean, data = xpnsl_recomb_df %>% filter(mean < 100)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0a731777-426c-4c16-8253-ff903289945a",
   "metadata": {},
   "outputs": [],
   "source": [
    "xpnsl_recomb_df %>% \n",
    "    filter(mean > 100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2c57f8ac-1cbd-42e2-bd7d-aaec28aaee26",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 8, repr.plot.height = 8)\n",
    "\n",
    "xpnsl_vs_recomb_plot <- xpnsl_recomb_df %>% \n",
    "    filter(mean < 100) %>% \n",
    "    ggplot(aes(x = mean, y = abs(xpnsl))) +\n",
    "        geom_point(size = 4) +\n",
    "        geom_smooth(method = \"lm\", color = \"black\", linewidth = 1.5) +\n",
    "        ylab(\"Normalized XP-nSL\") + xlab(\"Recombination rate (cM/Mb)\") +\n",
    "        theme_classic() +\n",
    "        theme(\n",
    "            legend.position = \"none\",\n",
    "            panel.border = element_blank(),\n",
    "            panel.grid.major.x = element_blank(),\n",
    "            panel.grid.minor.x = element_blank(),\n",
    "            axis.text = element_text(size=16),\n",
    "            axis.title = element_text(size=20)) +\n",
    "        transp_theme\n",
    "xpnsl_vs_recomb_plot\n",
    "ggsave(filename = snakemake@output[[\"xpnsl_vs_recomb_plot\"]], plot = xpnsl_vs_recomb_plot, \n",
    "       height = 8, width = 8, device = \"pdf\", dpi = 600, units = \"in\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "073d622d-04df-437e-a596-7bb0492538c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "head(gt_fst_df_filt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6013cf11-7d81-4963-a341-61752bde7e93",
   "metadata": {},
   "outputs": [],
   "source": [
    "fst_recomb_df <- gt_fst_df_filt %>% \n",
    "    filter(is_outlier == 1) %>% \n",
    "    dplyr::select(Chr, winCenter, avg_hudson_fst) %>% \n",
    "    left_join(., win_recomb, by = c(\"Chr\", \"winCenter\")) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1e6d69f3-c34e-4369-bbba-dc3f061ec12b",
   "metadata": {},
   "outputs": [],
   "source": [
    "summary(lm(avg_hudson_fst ~ mean, data = fst_recomb_df))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "175bacf6-9620-437a-95ef-7af1f6d56db9",
   "metadata": {},
   "outputs": [],
   "source": [
    "fst_vs_recomb_plot <- fst_recomb_df %>% \n",
    "    ggplot(aes(x = mean, y = avg_hudson_fst)) +\n",
    "        geom_point(size = 4) +\n",
    "        geom_smooth(method = \"lm\", color = \"black\", linewidth = 1.5) +\n",
    "        ylab(\"Hudson's Fst\") + xlab(\"Recombination rate (cM/Mb)\") +\n",
    "        theme_classic() +\n",
    "        theme(\n",
    "            legend.position = \"none\",\n",
    "            panel.border = element_blank(),\n",
    "            panel.grid.major.x = element_blank(),\n",
    "            panel.grid.minor.x = element_blank(),\n",
    "            axis.text = element_text(size=16),\n",
    "            axis.title = element_text(size=20)) +\n",
    "        transp_theme\n",
    "fst_vs_recomb_plot\n",
    "ggsave(filename = snakemake@output[[\"fst_vs_recomb_plot\"]], plot = fst_vs_recomb_plot, \n",
    "       height = 8, width = 8, device = \"pdf\", dpi = 600, units = \"in\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "98027e5b-ac52-4b9b-ab1e-0e9489acab11",
   "metadata": {},
   "source": [
    "# Create output tables\n",
    "\n",
    "## Dataframe with gene ID in top selected regions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c2cb12e4-9da8-4223-8bf9-79503592f704",
   "metadata": {},
   "outputs": [],
   "source": [
    "xpnsl_urban_outliers <- obs_xpnsl_ur_df_filt %>% filter(direction == \"Urban\") %>% mutate(id = paste0(Chr, '__', winCenter)) \n",
    "nsl_urban_outlier <- nsl_df_filt %>% filter(habitat == \"Urban\" & is_outlier == 1) %>% mutate(id = paste0(Chr, '__', winCenter))\n",
    "ihs_urban_outlier <- ihs_df_filt %>% filter(habitat == \"Urban\" & is_outlier == 1) %>% mutate(id = paste0(Chr, '__', winCenter))\n",
    "ihh12_urban_outlier <- ihh12_df_filt %>% filter(habitat == \"Urban\" & is_outlier == 1) %>% mutate(id = paste0(Chr, '__', winCenter))\n",
    "gt_fst_outlier <- gt_fst_df_filt %>% filter(is_outlier == 1) %>% mutate(id = paste0(Chr, '__', winCenter))\n",
    "td_outlier <- win_sfs_stats_filt %>% filter(is_outlier == 1) %>% mutate(id = paste0(Chr, '__', winCenter))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "834fda2a-76ab-42b2-afb5-26bc35f945d5",
   "metadata": {},
   "outputs": [],
   "source": [
    "xpnsl_rural_outliers <- obs_xpnsl_ur_df_filt %>% filter(direction == \"Rural\") %>% mutate(id = paste0(Chr, '__', winCenter))\n",
    "nsl_rural_outlier <- nsl_df_filt %>% filter(habitat == \"Rural\" & is_outlier == 1) %>% mutate(id = paste0(Chr, '__', winCenter))\n",
    "ihs_rural_outlier <- ihs_df_filt %>% filter(habitat == \"Rural\" & is_outlier == 1) %>% mutate(id = paste0(Chr, '__', winCenter))\n",
    "ihh12_rural_outlier <- ihh12_df_filt %>% filter(habitat == \"Rural\" & is_outlier == 1) %>% mutate(id = paste0(Chr, '__', winCenter))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5b680337-5252-495c-91cb-cbf30c2144cb",
   "metadata": {},
   "outputs": [],
   "source": [
    "get_positions <- function(df){\n",
    "    df_out <- df %>% \n",
    "        dplyr::select(Chr, start, end)\n",
    "    return(df_out)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "61e4793a-b0d5-4fb5-8b65-bed5951a825d",
   "metadata": {},
   "outputs": [],
   "source": [
    "xpnsl_topHits_positions <- get_positions(xpnsl_topHits_urban_rural)\n",
    "nsl_urban_positions <- get_positions(nsl_urban_outlier)\n",
    "nsl_rural_positions <- get_positions(nsl_rural_outlier)\n",
    "ihs_urban_positions <- get_positions(ihs_urban_outlier)\n",
    "ihs_rural_positions <- get_positions(ihs_rural_outlier)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6a731ba0-9cc5-4368-9d3d-0ced8b49de1c",
   "metadata": {},
   "outputs": [],
   "source": [
    "xpnsl_topHits_urban_rural"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fe0bdc68-ede6-4cd8-b6a4-d71149486c9b",
   "metadata": {},
   "outputs": [],
   "source": [
    "tajima_vals <- bt.intersect(xpnsl_topHits_positions, td_outlier, wa = T, wb = T) %>% \n",
    "    dplyr::select(V1, V2, V3, V10, V11)\n",
    "names(tajima_vals) <- c(\"Chr\", \"start\", \"end\", \"Tajima_Urban\", \"Tajima_Rural\")\n",
    "tajima_vals <- tajima_vals %>% \n",
    "    group_by(Chr, start, end) %>% \n",
    "    summarise(Tajima_Urban = mean(Tajima_Urban),\n",
    "              Tajima_Rural = mean(Tajima_Rural))\n",
    "tajima_vals"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f115b30c-8adc-4b5b-aa42-0ad7eb7d4b06",
   "metadata": {},
   "outputs": [],
   "source": [
    "count_overlapping_outlier_windows <- function(pos1, pos2){\n",
    "    \n",
    "    # Count number of overlapping statistic outlier windows in each XP-nSL window\n",
    "    num_outliers <- bt.intersect(pos1, pos2, c = T) %>% pull(V4)\n",
    "    return(num_outliers)\n",
    "} "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "de8693b7-4d83-4300-9b10-6d9788092c0f",
   "metadata": {},
   "outputs": [],
   "source": [
    "xpnsl_topHits_urban_rural_with_other_stats <- xpnsl_topHits_urban_rural %>% \n",
    "    mutate(num_nsl_urban = count_overlapping_outlier_windows(get_positions(.), get_positions(nsl_urban_outlier)),\n",
    "           num_nsl_rural = count_overlapping_outlier_windows(get_positions(.), get_positions(nsl_rural_outlier)),\n",
    "           num_ihs_urban = count_overlapping_outlier_windows(get_positions(.), get_positions(ihs_urban_outlier)),\n",
    "           num_ihs_rural = count_overlapping_outlier_windows(get_positions(.), get_positions(ihs_rural_outlier)),\n",
    "           num_ihh12_urban = count_overlapping_outlier_windows(get_positions(.), get_positions(ihh12_urban_outlier)),\n",
    "           num_ihh12_rural = count_overlapping_outlier_windows(get_positions(.), get_positions(ihh12_rural_outlier)),\n",
    "           num_fst = count_overlapping_outlier_windows(get_positions(.), get_positions(gt_fst_outlier)),\n",
    "           num_td = count_overlapping_outlier_windows(get_positions(.), get_positions(td_outlier))) %>% \n",
    "    left_join(., tajima_vals, by = c(\"Chr\", \"start\", \"end\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8333d774-1157-428c-b76b-bb781466387e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load GFF\n",
    "gff <- ape::read.gff(snakemake@input[['gff']], GFF3 = TRUE) %>% \n",
    "    dplyr::select(seqid, start, end, everything())\n",
    "\n",
    "# Get gene names dataframe\n",
    "gene_names <- gff %>% \n",
    "    filter(type == 'gene') %>% \n",
    "    mutate(gene = str_extract(attributes, pattern = '(?<=gene=)\\\\w+(?=;)'),\n",
    "           gene_id = str_extract(attributes, pattern = '(?<=ID\\\\=)ACLI19_g\\\\d+(?=;)')) %>% \n",
    "    dplyr::select(seqid, start, end, gene_id, gene) %>% \n",
    "    rename('Chr' = 'seqid') %>% \n",
    "    remap_chr_names()\n",
    "\n",
    "# Get gene products and GO annotations dataframe\n",
    "prod_go_annot <- gff %>% \n",
    "    filter(type == 'mRNA') %>% \n",
    "    mutate(id = str_extract(attributes, pattern = '(?<=ID\\\\=)ACLI19_g\\\\d+\\\\.t\\\\d+(?=;)'),\n",
    "           func = str_extract(attributes, pattern = '(?<=product=)[^;]*'),\n",
    "           go = str_extract_all(attributes, pattern = 'GO:\\\\d+(?=(,|;))')) %>% \n",
    "    separate(id, into = c('gene_id', 'trans'), sep = '\\\\.') %>% \n",
    "    filter(trans == 't1') %>% \n",
    "    dplyr::select(seqid, gene_id, func, go) %>% \n",
    "    rename('Chr' = 'seqid') %>% \n",
    "    remap_chr_names()\n",
    "\n",
    "# Combine genes, functions, and GO annotations into single dataframe\n",
    "genes_prods_go_df <- left_join(gene_names, prod_go_annot, by = c('Chr', 'gene_id'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c8aee512-2323-45d7-a792-ab47b9b65813",
   "metadata": {},
   "outputs": [],
   "source": [
    "xpnsl_topHits_urban_rural_with_genes_long <- bt.intersect(xpnsl_topHits_urban_rural %>% \n",
    "                                                            dplyr::select(Chr, start, end, direction),\n",
    "                                                        genes_prods_go_df %>%   \n",
    "                                                            dplyr::select(-go),     \n",
    "                                                        wa=T, wb=T) %>%           \n",
    "    dplyr::select(V1, V2, V3, V4, V8, V9, V10)                                    \n",
    "names(xpnsl_topHits_urban_rural_with_genes_long) <- c('Chr', 'start', 'end', 'direction', 'gene_id', 'gene_symbol', 'product')\n",
    "head(xpnsl_topHits_urban_rural_with_genes_long)\n",
    "\n",
    "# Write dataframe with gene IDs for top ten selected urban and rural regions\n",
    "write_delim(xpnsl_topHits_urban_rural_with_genes_long, snakemake@output[['top_hits_genes']], delim = '\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dcb15685-0117-4ec3-8ffb-8e403e8d1185",
   "metadata": {},
   "source": [
    "## Table with gene symbols, products, and Fst for top selected regions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "368543e8-a300-4c46-83bd-f9a769d3b5f1",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Concatenate gene symbols for top 10 selected regions\n",
    "symbols <- xpnsl_topHits_urban_rural_with_genes_long %>% \n",
    "    dplyr::select(-gene_id, -product) %>% \n",
    "    filter(!is.na(gene_symbol)) %>% \n",
    "    group_by(Chr, start, end, direction) %>% \n",
    "    summarise(gene_symbols = toString(gene_symbol))\n",
    "\n",
    "# Concatenate products for top 10 selected regions\n",
    "prods <- xpnsl_topHits_urban_rural_with_genes_long %>% \n",
    "    dplyr::select(-gene_id, -gene_symbol) %>% \n",
    "    filter(product != 'hypothetical protein') %>% \n",
    "    group_by(Chr, start, end, direction) %>% \n",
    "    summarise(products = toString(product))\n",
    "\n",
    "# Add symbols and products to table\n",
    "xpnsl_topHits_urban_rural_with_genes <- xpnsl_topHits_urban_rural_with_other_stats %>% \n",
    "    left_join(., symbols) %>% \n",
    "    left_join(., prods)\n",
    "\n",
    "xpnsl_topHits_urban_rural_with_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7856b4ec-3f94-4c74-b3a4-c9fd9fc8f74d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write table with gene symbols, products, and Fst overlaps for top ten selected yrban and rural regions\n",
    "write_delim(xpnsl_topHits_urban_rural_with_genes, snakemake@output[['top_hits_tbl']], delim = '\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "57e925bb-47ce-4696-a3a4-2ba91716bb36",
   "metadata": {},
   "source": [
    "## Dataframe with gene IDs in selected regions based on XP-nSL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9132948e-5385-4591-94cb-07b5f952df3a",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# All selected regions\n",
    "xpnsl_outliers <- obs_xpnsl_ur_df_filt %>% \n",
    "    filter(direction != 'None') %>% \n",
    "    dplyr::select(Chr, start, end, direction)\n",
    "\n",
    "xpnsl_outliers_with_genes <- bt.intersect(xpnsl_outliers, gene_names, wa = T, wb = T) %>% \n",
    "    dplyr::select(V1, V2, V3, V4, V8)\n",
    "names(xpnsl_outliers_with_genes) <- c('Chr', 'start', 'end', 'direction', 'gene_id')\n",
    "xpnsl_outliers_with_genes\n",
    "\n",
    "# Write dataframe with gene IDs for all selected urban and rural regions\n",
    "write_delim(xpnsl_outliers_with_genes, snakemake@output[['xpnsl_out_genes']], delim = '\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b1f7ec6d-037a-4877-a606-cc5aea0d254d",
   "metadata": {},
   "source": [
    "## Polygenic selection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0cb857b8-207b-4b25-975f-49385d1c4bdd",
   "metadata": {},
   "outputs": [],
   "source": [
    "load_persite_xpnsl <- function(path){\n",
    "    base <- basename(path)\n",
    "    chrom <- str_extract(string = base, pattern = \".*(?=_Urban)\")\n",
    "    df_out <- suppressMessages(read_delim(path, delim=\"\\t\")) %>% \n",
    "        mutate(Chr = chrom) %>% \n",
    "    remap_chr_names()\n",
    "    return(df_out)\n",
    "}\n",
    "xpnsl_persite <- snakemake@input[[\"norm_xpnsl\"]] %>% \n",
    "    purrr::map_dfr(., load_persite_xpnsl) %>% \n",
    "    rename(\"normxpnsl\" = \"normxpehh\") %>% \n",
    "    mutate(crit = ifelse(normxpnsl < -2 | normxpnsl > 2, 1, 0),\n",
    "           direction = case_when(crit == 1 & normxpnsl > 0 ~ \"Urban\",\n",
    "                                 crit == 1 & normxpnsl < 0 ~ \"Rural\",\n",
    "                                 TRUE ~ \"None\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fac7c295-b9b6-4d48-b636-009b3405a685",
   "metadata": {},
   "outputs": [],
   "source": [
    "num_genes_df <- gene_names %>% \n",
    "    group_by(Chr) %>% \n",
    "    summarise(num_genes = n())\n",
    "\n",
    "chr_lengths <- read_delim(snakemake@input[[\"chr_lengths\"]], delim=\"\\t\", \n",
    "                          col_names = c(\"Chr\", \"size\")) %>% \n",
    "    remap_chr_names()\n",
    "\n",
    "num_xpnsl_outliers_by_chr_urban <- xpnsl_persite %>% \n",
    "    filter(direction == \"Urban\") %>% \n",
    "    group_by(Chr) %>% \n",
    "    summarise(Urban = n())\n",
    "\n",
    "num_xpnsl_outliers_by_chr_rural <- xpnsl_persite %>% \n",
    "    filter(direction == \"Rural\") %>% \n",
    "    group_by(Chr) %>% \n",
    "    summarise(Rural = n())\n",
    "\n",
    "chr_size_outliers_df <- chr_lengths %>% \n",
    "     left_join(., num_genes_df, by = \"Chr\") %>% \n",
    "     left_join(., num_xpnsl_outliers_by_chr_urban, by = \"Chr\") %>% \n",
    "     left_join(., num_xpnsl_outliers_by_chr_rural, by = \"Chr\") %>% \n",
    "     filter(!is.na(Chr)) %>% \n",
    "     pivot_longer(Urban:Rural, names_to = \"habitat\", values_to = \"num_outliers\")\n",
    "\n",
    "head(chr_size_outliers_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "24aee5ae-5e3d-4ee9-989b-f2b5228c9e1a",
   "metadata": {},
   "outputs": [],
   "source": [
    "summary(lm(num_outliers ~ num_genes*habitat, data = chr_size_outliers_df))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5e6f8cb4-1c1e-4eba-9dc0-f49d0904f5b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 8, repr.plot.height = 8)\n",
    "num_xpnsl_by_num_genes <- chr_size_outliers_df %>% \n",
    "    ggplot(aes(x = num_genes, y = num_outliers)) + \n",
    "        geom_point(size = 4, aes(color = habitat, fill = habitat)) +\n",
    "        geom_smooth(method = \"lm\", linewidth = 1.5,\n",
    "                    aes(color = habitat)) +\n",
    "        ylab(\"Number of XP-nSL outliers\") + xlab(\"Number of genes\") +\n",
    "        scale_color_manual(values = c(\"#007243\", \"#003876\")) +\n",
    "        scale_fill_manual(values = c(\"#007243\", \"#003876\")) +\n",
    "        theme_classic() +\n",
    "        theme(\n",
    "            legend.position = \"none\",\n",
    "            panel.border = element_blank(),\n",
    "            panel.grid.major.x = element_blank(),\n",
    "            panel.grid.minor.x = element_blank(),\n",
    "            axis.text = element_text(size=16),\n",
    "            axis.title = element_text(size=20)) +\n",
    "        transp_theme\n",
    "num_xpnsl_by_num_genes\n",
    "\n",
    "ggsave(filename = snakemake@output[[\"num_xpnsl_by_num_genes\"]], plot = num_xpnsl_by_num_genes, \n",
    "       height = 8, width = 8, device = \"pdf\", dpi = 600, units = \"in\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "54d14d6f-d366-4746-b446-15567d757885",
   "metadata": {},
   "source": [
    "## Zoomed-in Manhattan and Haplotype plots\n",
    "\n",
    "### Positive selection in urban habitats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9a7bbe6b-63d2-47a4-b00b-d8994b2f9fb3",
   "metadata": {},
   "outputs": [],
   "source": [
    "xpnsl_topHits_urban_rural_with_other_stats %>% \n",
    "    filter(direction == 'Urban') %>% \n",
    "    dplyr::select(Chr, start, end, win_size, min_xpnsl_rank, num_nsl_urban:num_fst)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9ac2531b-b9f8-4ca6-9eb9-ad2f69188e0b",
   "metadata": {},
   "outputs": [],
   "source": [
    "max_xpnsl <- xpnsl_persite %>% filter(normxpnsl == max(normxpnsl))\n",
    "max_xpnsl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a96bd2e5-95c3-4a25-8910-2cd77ae4e24b",
   "metadata": {},
   "outputs": [],
   "source": [
    "max <- max_xpnsl %>% dplyr::select(Chr, pos) %>% rename('start' = 'pos') %>% mutate(end = start) %>% dplyr::select(Chr, start, end)\n",
    "genes <- genes_prods_go_df %>% dplyr::select(Chr, start, end, gene, func)\n",
    "bt.intersect(max, genes, wb=T)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7e46cd50-f92d-4475-a27b-4a7d1f8ac019",
   "metadata": {},
   "source": [
    "#### Chromosome 7"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ae37a0a8-2562-488a-86ba-0af8fe1f2698",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function for formatting x-axis numbers\n",
    "formatter1e6 <- function(x){ \n",
    "    x <- x / 1e6\n",
    "    return(sprintf(x, fmt = '%#.2f'))\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6e3a5d56-6fa8-4ce5-ba8a-dddf92fb99c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "region_df <- xpnsl_topHits_urban_rural %>% \n",
    "    filter(direction == 'Urban') %>% \n",
    "    filter(Chr == 7) %>%\n",
    "    dplyr::select(Chr, start, end)\n",
    "\n",
    "chrom <- region_df %>% pull(Chr) %>% unique()\n",
    "min_pos <- region_df %>% pull(start) %>% min()\n",
    "max_pos <- region_df %>% pull(end) %>% max()\n",
    "buffer <- 500000\n",
    "\n",
    "all_outlier_regions <- xpnsl_urban_outliers %>% \n",
    "    filter(Chr == chrom & start >= min_pos - buffer & end <= max_pos + buffer) %>% \n",
    "    dplyr::select(Chr, start, end)\n",
    "\n",
    "region_raw_df <- xpnsl_persite %>%\n",
    "    filter(Chr == chrom & pos >= min_pos - buffer & pos <= max_pos + buffer)\n",
    "\n",
    "xpnsl_not_outlier <- region_raw_df %>% filter(crit == 0)\n",
    "urban_xpnsl_outliers <- region_raw_df %>% filter(crit == 1 & normxpnsl > 0)\n",
    "rural_xpnsl_outliers <- region_raw_df %>% filter(crit == 1 & normxpnsl < 0)\n",
    "\n",
    "fst_outliers <- gt_fst_df_filt %>% filter(is_outlier == 1) %>% dplyr::select(Chr, start, end)\n",
    "fst_wins <- bt.intersect(fst_outliers, all_outlier_regions, wa = T) %>% \n",
    "    rename('Chr' = 'V1', 'start' = 'V2', 'end' = 'V3')\n",
    "nsl_outliers <- nsl_df_filt %>% filter(is_outlier == 1) %>% dplyr::select(Chr, start, end)\n",
    "nsl_wins <- bt.intersect(nsl_outliers, all_outlier_regions, wa = T) %>% \n",
    "    rename('Chr' = 'V1', 'start' = 'V2', 'end' = 'V3')\n",
    "\n",
    "manhat_plot <- ggplot(xpnsl_not_outlier, aes(x = pos, y = normxpnsl)) +\n",
    "    geom_rect(data=region_df, aes(xmin=start, xmax=end), ymin=-Inf, ymax=Inf, fill=\"black\", alpha=0.3, inherit.aes = F) +\n",
    "    geom_rect(data=all_outlier_regions, aes(xmin=start, xmax=end), ymin=-Inf, ymax=Inf, fill=\"black\", alpha=0.1, inherit.aes = F) +\n",
    "    geom_point(data = xpnsl_not_outlier, shape = 21, alpha = 0.3, size = 1, , fill = \"black\", color = \"black\") +\n",
    "    geom_point(data = urban_xpnsl_outliers, shape = 21, alpha = 0.3, size = 1, color = '#003876', fill = '#003876',\n",
    "               aes(x = pos, y = normxpnsl)) +\n",
    "    geom_point(data = rural_xpnsl_outliers, shape = 21, alpha = 0.3, size = 1, color = '#007243', fill = '#007243',\n",
    "               aes(x = pos, y = normxpnsl)) +\n",
    "    geom_hline(yintercept = 2, color = \"grey40\", linetype = \"dashed\") +\n",
    "    geom_hline(yintercept = -2, color = \"grey40\", linetype = \"dashed\") +\n",
    "    xlab('Chromosome 7 (position in Mbp)') + ylab('Normalized XP-nSL') +\n",
    "    coord_cartesian(ylim = c(-3.25, 6.5)) +\n",
    "    geom_segment(data = fst_wins, aes(x = start, xend = end), y = -3, yend = -3, color = '#d62828', linewidth = 1) +\n",
    "    geom_segment(data = nsl_wins, aes(x = start, xend = end), y = -3.25, yend = -3.25, color = '#ffbe0b', linewidth = 1) +\n",
    "    scale_x_continuous(breaks = seq(min_pos - buffer, max_pos + buffer, 500000), labels = formatter1e6) +\n",
    "    theme_classic() +\n",
    "    theme(\n",
    "        panel.border = element_blank(),\n",
    "        panel.grid.major.x = element_blank(),\n",
    "        panel.grid.minor.x = element_blank(),\n",
    "        axis.text = element_text(size=16),\n",
    "        axis.title = element_text(size=20),\n",
    "      ) +\n",
    "    transp_theme\n",
    "\n",
    "options(repr.plot.width = 8, repr.plot.height = 6)\n",
    "manhat_plot\n",
    "ggsave(filename = snakemake@output[[\"Chr04_Occ_urb_xpnsl\"]], plot = manhat_plot, \n",
    "       height = 6, width = 8, device = \"pdf\", dpi = 600, units = \"in\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fd9cf99a-1a4f-432c-878e-5240f3ba955d",
   "metadata": {},
   "outputs": [],
   "source": [
    "chr04_Occ_vcf <- vcfR::read.vcfR(snakemake@input[[\"vcfs\"]][7])\n",
    "popmap <- read_delim(snakemake@input[[\"popmap\"]], delim = \"\\t\", col_names = c(\"ind\", \"pop\")) %>% \n",
    "    filter(pop != \"Suburban\") %>% \n",
    "    as.data.frame()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e92b3c00-fdff-474d-9e9c-ed7fb2717f24",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Subset VCF around same region as above\n",
    "idx <- which(vcfR::getPOS(chr04_Occ_vcf) >= min_pos - buffer & vcfR::getPOS(chr04_Occ_vcf) <= max_pos + buffer)\n",
    "start_idx <- head(idx, n = 1)\n",
    "end_idx <- tail(idx, n = 1)\n",
    "chr04_Occ_vcf_urb_sub <- chr04_Occ_vcf[start_idx:end_idx]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5a70cdf2-7616-49b8-895f-d628b8344b8c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create genotype plot and cluster genotypes with PCA and hclust\n",
    "chr04_Occ_urb_clustered <- genotype_plot(vcf_object = chr04_Occ_vcf_urb_sub, popmap = popmap, snp_label_size = 500000, cluster = T)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b70b7a7b-5ee4-40ff-a139-3d5007363409",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Reorder popmap based on clustering. This is only for visualization\n",
    "popmap2 <- data.frame(ind = chr04_Occ_urb_clustered$dendro_labels) %>% \n",
    "    left_join(popmap, by = \"ind\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f4d0e33e-601d-46cc-9621-4fc72c673ff6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create final haplotype plot with colors. Unclustered and split genotypes into phased haplotypes\n",
    "chr04_Occ_urb_final <- genotype_plot(vcf_object = chr04_Occ_vcf_urb_sub, \n",
    "                                     popmap = popmap2, snp_label_size = 500000, \n",
    "                                     cluster = F, plot_phased = T)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "874ff40f-bd54-45d4-91e3-494163aa554d",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 20, repr.plot.height = 6)\n",
    "Chr04_Occ_urb_ur_haps <- chr04_Occ_urb_final$genotypes + \n",
    "    geom_rect(data=region_df, aes(xmin=start, xmax=end), ymin=0, ymax=164.5, fill=\"black\", alpha=0.3, inherit.aes = F) +\n",
    "    geom_rect(data=all_outlier_regions, aes(xmin=start, xmax=end), ymin=0, ymax=164.5, fill=\"black\", alpha=0.1, inherit.aes = F) +\n",
    "    scale_x_continuous(expand = c(0,0), limit = c(min_pos - buffer, max_pos + buffer), \n",
    "                       breaks = seq(min_pos - buffer, max_pos + buffer, 500000), labels = formatter1e6) +\n",
    "    scale_fill_manual(values = c(\"#fcbf49\", \"#d62828\"), name=\"Haplotype\",\n",
    "                      breaks=c(\"0\",\"1\"),labels=c(\"REF\",\"ALT\")) +\n",
    "    theme(\n",
    "        axis.text.x = element_text(size=16),\n",
    "        axis.title = element_text(size=20),\n",
    "        axis.text.y = element_text(angle = 90),\n",
    "        axis.ticks.x = element_line(),\n",
    "        legend.position = \"top\"\n",
    "      ) +\n",
    "    transp_theme\n",
    "Chr04_Occ_urb_ur_haps\n",
    "ggsave(filename = snakemake@output[[\"Chr04_Occ_urb_ur_haps\"]], plot = Chr04_Occ_urb_ur_haps, \n",
    "       height = 8, width = 20, device = \"pdf\", dpi = 600, units = \"in\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cfcf612c-d65f-47f5-937e-a5756dd0f1b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "subset_vcf <- function(df, vcf){\n",
    "    start <- df %>% pull(start)\n",
    "    end <- df %>% pull(end)\n",
    "    idx <- which(vcfR::getPOS(vcf) >= start & vcfR::getPOS(vcf) <= end)\n",
    "    start_idx <- head(idx, n = 1)\n",
    "    end_idx <- tail(idx, n = 1)\n",
    "    vcf_sub <- vcf[start_idx:end_idx]\n",
    "    return(vcf_sub)\n",
    "}\n",
    "\n",
    "all_vcfs <- all_outlier_regions %>% \n",
    "    group_split(start) %>% \n",
    "    purrr::map(., subset_vcf, vcf = chr04_Occ_vcf_urb_sub)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "859172d8-5864-463e-a35d-4d4d98531ba9",
   "metadata": {},
   "outputs": [],
   "source": [
    "combined_vcf <- do.call(\"rbind\", all_vcfs)\n",
    "chr04_Occ_urb_sel_AF <- genotype_plot(vcf_object = combined_vcf, popmap = popmap, snp_label_size = 500000, \n",
    "                                      plot_allele_frequency = T)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fedc7850-8e7e-4ae1-a8f2-e2d15cc03b53",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 8, repr.plot.height = 6)\n",
    "cols_hab <- c(\"#007243\", \"red\", \"#003876\")\n",
    "Chr04_Occ_urb_ur_af <- chr04_Occ_urb_sel_AF$genotypes$data %>% \n",
    "    ggplot(aes(x = AF)) +\n",
    "        geom_density(aes(color = pop, fill = pop), alpha = 0.7) +\n",
    "        scale_color_manual(values = cols_hab) +\n",
    "        scale_fill_manual(values = cols_hab) +\n",
    "        ylab(\"Density\") + \n",
    "        xlab(\"Frequency of REF allele\") +\n",
    "        theme_classic() +\n",
    "        theme(axis.text = element_text(size = 18),\n",
    "              axis.title = element_text(size = 20),\n",
    "              legend.position = 'top',\n",
    "              legend.title = element_text(size = 16),\n",
    "              legend.text = element_text(size = 14)) +\n",
    "        transp_theme\n",
    "Chr04_Occ_urb_ur_af\n",
    "ggsave(filename = snakemake@output[[\"Chr04_Occ_urb_ur_af\"]], plot = Chr04_Occ_urb_ur_af, \n",
    "       height = 6, width = 8, device = \"pdf\", dpi = 600, units = \"in\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d4b3fcc8-ba3a-46cb-ac8e-f5683a32cfbe",
   "metadata": {},
   "outputs": [],
   "source": [
    "chr04_Occ_urb_sel_clust <- genotype_plot(vcf_object = combined_vcf, popmap = popmap, snp_label_size = 500000, \n",
    "                                         cluster = T, plot_phased = T)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "09793040-26ed-4e93-8619-6465ada61178",
   "metadata": {},
   "outputs": [],
   "source": [
    "var_PC1 <- round(get_eigenvalue(chr04_Occ_urb_sel_clust$cluster_pca)[\"Dim.1\",\"variance.percent\"], 1)\n",
    "var_PC2 <- round(get_eigenvalue(chr04_Occ_urb_sel_clust$cluster_pca)[\"Dim.2\",\"variance.percent\"], 1)\n",
    "\n",
    "Chr04_Occ_urb_ur_pca <- chr04_Occ_urb_sel_clust$cluster_pca$l1 %>% \n",
    "    as.data.frame() %>% \n",
    "    rownames_to_column(., var = \"tmp\") %>% \n",
    "    mutate(ind = str_sub(tmp, 1, -3),\n",
    "           hap = str_sub(tmp, -1)) %>% \n",
    "    dplyr::select(ind, hap, RS1, RS2) %>% \n",
    "    left_join(popmap, by = \"ind\") %>% \n",
    "    ggplot(., aes(x = RS1, y = RS2)) +\n",
    "            geom_point(aes(color = pop, shape = pop), size = 7, alpha = 0.75) +\n",
    "            scale_color_manual(values = cols_hab) +\n",
    "            theme_classic() +\n",
    "            xlab(paste0('PC1 (', var_PC1, '%)')) + ylab(paste0('PC2 (', var_PC2, '%)')) +\n",
    "            theme(axis.text = element_text(size = 18),\n",
    "                  axis.title = element_text(size = 20),\n",
    "                  legend.position = 'top',\n",
    "                  legend.title = element_text(size = 16),\n",
    "                  legend.text = element_text(size = 14)) +\n",
    "    transp_theme\n",
    "Chr04_Occ_urb_ur_pca\n",
    "ggsave(filename = snakemake@output[[\"Chr04_Occ_urb_ur_pca\"]], plot = Chr04_Occ_urb_ur_pca, \n",
    "       height = 6, width = 8, device = \"pdf\", dpi = 600, units = \"in\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a868c422-ca4b-46bb-87dd-bd8ed981b7b1",
   "metadata": {},
   "source": [
    "#### Chromosome 9"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2a5c749a-7296-4f96-a725-cfbce69cb564",
   "metadata": {},
   "outputs": [],
   "source": [
    "region_df <- xpnsl_topHits_urban_rural %>% \n",
    "    filter(direction == 'Urban') %>% \n",
    "    filter(Chr == 9) %>%\n",
    "    dplyr::select(Chr, start, end)\n",
    "\n",
    "chrom <- region_df %>% pull(Chr) %>% unique()\n",
    "min_pos <- region_df %>% pull(start) %>% min()\n",
    "max_pos <- region_df %>% pull(end) %>% max()\n",
    "buffer <- 500000\n",
    "\n",
    "all_outlier_regions <- xpnsl_urban_outliers %>% \n",
    "    filter(Chr == chrom & start >= min_pos - buffer & end <= max_pos + buffer) %>% \n",
    "    dplyr::select(Chr, start, end)\n",
    "\n",
    "region_raw_df <- xpnsl_persite %>%\n",
    "    filter(Chr == chrom & pos >= min_pos - buffer & pos <= max_pos + buffer)\n",
    "\n",
    "xpnsl_not_outlier <- region_raw_df %>% filter(crit == 0)\n",
    "urban_xpnsl_outliers <- region_raw_df %>% filter(crit == 1 & normxpnsl > 0)\n",
    "rural_xpnsl_outliers <- region_raw_df %>% filter(crit == 1 & normxpnsl < 0)\n",
    "\n",
    "formatter1e6 <- function(x){ \n",
    "    x <- x / 1e6\n",
    "    return(sprintf(x, fmt = '%#.2f'))\n",
    "}\n",
    "\n",
    "fst_outliers <- gt_fst_df_filt %>% filter(is_outlier == 1) %>% dplyr::select(Chr, start, end)\n",
    "fst_wins <- bt.intersect(fst_outliers, all_outlier_regions, wa = T) %>% \n",
    "    rename('Chr' = 'V1', 'start' = 'V2', 'end' = 'V3')\n",
    "nsl_outliers <- nsl_df_filt %>% filter(is_outlier == 1 & habitat == \"Urban\") %>% dplyr::select(Chr, start, end)\n",
    "nsl_wins <- bt.intersect(nsl_outliers, all_outlier_regions, wa = T) %>% \n",
    "    rename('Chr' = 'V1', 'start' = 'V2', 'end' = 'V3')\n",
    "ihs_outliers <- ihs_df_filt %>% filter(is_outlier == 1 & habitat == \"Urban\") %>% dplyr::select(Chr, start, end)\n",
    "ihs_wins <- bt.intersect(ihs_outliers, all_outlier_regions, wa = T) %>% \n",
    "    rename('Chr' = 'V1', 'start' = 'V2', 'end' = 'V3')\n",
    "ihh12_outliers <- ihh12_df_filt %>% filter(is_outlier == 1 & habitat == \"Urban\") %>% dplyr::select(Chr, start, end)\n",
    "ihh12_wins <- bt.intersect(ihh12_outliers, all_outlier_regions, wa = T) %>% \n",
    "    rename('Chr' = 'V1', 'start' = 'V2', 'end' = 'V3')\n",
    "\n",
    "manhat_plot <- ggplot(xpnsl_not_outlier, aes(x = pos, y = normxpnsl)) +\n",
    "    geom_rect(data=region_df, aes(xmin=start, xmax=end), ymin=-Inf, ymax=Inf, fill=\"black\", alpha=0.3, inherit.aes = F) +\n",
    "    geom_rect(data=all_outlier_regions, aes(xmin=start, xmax=end), ymin=-Inf, ymax=Inf, fill=\"black\", alpha=0.1, inherit.aes = F) +   \n",
    "    geom_point(shape = 21, alpha = 0.3, size = 1, , fill = \"black\", color = \"black\") +\n",
    "    geom_point(data = urban_xpnsl_outliers, shape = 21, alpha = 0.3, size = 1, color = '#003876', fill = '#003876',\n",
    "               aes(x = pos, y = normxpnsl)) +\n",
    "    geom_point(data = rural_xpnsl_outliers, shape = 21, alpha = 0.3, size = 1, color = '#007243', fill = '#007243',\n",
    "               aes(x = pos, y = normxpnsl)) +\n",
    "    geom_point(data = max_xpnsl, shape = 23, size = 3.5, fill = 'yellow', alpha = 1) +\n",
    "    geom_hline(yintercept = 2, color = \"grey40\", linetype = \"dashed\") +\n",
    "    geom_hline(yintercept = -2, color = \"grey40\", linetype = \"dashed\") +\n",
    "    xlab('Chromosome 9 (position in Mbp)') + ylab('Normalized XP-nSL') +\n",
    "    coord_cartesian(ylim = c(-3.5, 6.5)) +\n",
    "    geom_segment(data = fst_wins, aes(x = start, xend = end), y = -3, yend = -3, color = '#d62828', linewidth = 1) +\n",
    "    geom_segment(data = nsl_wins, aes(x = start, xend = end), y = -3.25, yend = -3.25, color = '#ffbe0b', linewidth = 1) +\n",
    "    geom_segment(data = ihs_wins, aes(x = start, xend = end), y = -3.5, yend = -3.5, color = '#8338ec', linewidth = 1) +\n",
    "    geom_segment(data = ihh12_wins, aes(x = start, xend = end), y = -3.75, yend = -3.75, color = '#386641', linewidth = 1) +\n",
    "    scale_x_continuous(breaks = seq(min_pos - buffer, max_pos + buffer, 200000), labels = formatter1e6) +\n",
    "    # scale_y_continuous(breaks = seq(-2, 6, 2)) +\n",
    "    theme_classic() +\n",
    "    theme(\n",
    "        panel.border = element_blank(),\n",
    "        panel.grid.major.x = element_blank(),\n",
    "        panel.grid.minor.x = element_blank(),\n",
    "        axis.text = element_text(size=16),\n",
    "        axis.title = element_text(size=20),\n",
    "      ) +\n",
    "    transp_theme\n",
    "\n",
    "options(repr.plot.width = 8, repr.plot.height = 6)\n",
    "manhat_plot\n",
    "ggsave(filename = snakemake@output[[\"Chr05_Occ_urb_xpnsl\"]], plot = manhat_plot, \n",
    "       height = 6, width = 8, device = \"pdf\", dpi = 600, units = \"in\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "857886c2-53ad-4e55-84c6-c5e1b46ffd9c",
   "metadata": {},
   "outputs": [],
   "source": [
    "chr05_Occ_vcf <- vcfR::read.vcfR(snakemake@input[[\"vcfs\"]][9])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4e15f0e5-26aa-467a-92fc-e6819aca6109",
   "metadata": {},
   "outputs": [],
   "source": [
    "idx <- which(vcfR::getPOS(chr05_Occ_vcf) >= min_pos - buffer & vcfR::getPOS(chr05_Occ_vcf) <= max_pos + buffer)\n",
    "start_idx <- head(idx, n = 1)\n",
    "end_idx <- tail(idx, n = 1)\n",
    "chr05_Occ_vcf_urb_sub <- chr05_Occ_vcf[start_idx:end_idx]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c3d82364-f06c-442e-9e91-efee8d0570e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create genotype plot and cluster genotypes with PCA and hclust\n",
    "chr05_Occ_urb_clustered <- genotype_plot(vcf_object = chr05_Occ_vcf_urb_sub, popmap = popmap, snp_label_size = 200000, cluster = T)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5f9d6786-8db6-4a20-873c-ef0b1a0890b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Reorder popmap based on clustering. This is only for visualization\n",
    "popmap2 <- data.frame(ind = chr05_Occ_urb_clustered$dendro_labels) %>% \n",
    "    left_join(popmap, by = \"ind\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "995a3542-859b-41bc-bcc3-760136b8efe6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create final haplotype plot with colors. Unclustered and split genotypes into phased haplotypes\n",
    "chr05_Occ_urb_final <- genotype_plot(vcf_object = chr05_Occ_vcf_urb_sub, \n",
    "                                     popmap = popmap2, snp_label_size = 200000, \n",
    "                                     cluster = F, plot_phased = T)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0b259802-7b03-4cc5-9c75-493779caaa14",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 20, repr.plot.height = 6)\n",
    "Chr05_Occ_urb_ur_haps <- chr05_Occ_urb_final$genotypes + \n",
    "    geom_rect(data=region_df, aes(xmin=start, xmax=end), ymin=0, ymax=164.5, fill=\"black\", alpha=0.3, inherit.aes = F) +\n",
    "    geom_rect(data=all_outlier_regions, aes(xmin=start, xmax=end), ymin=0, ymax=164.5, fill=\"black\", alpha=0.1, inherit.aes = F) +\n",
    "    scale_x_continuous(expand = c(0,0), limit = c(min_pos - buffer, max_pos + buffer), \n",
    "                       breaks = seq(min_pos - buffer, max_pos + buffer, 200000), labels = formatter1e6) +\n",
    "    scale_fill_manual(values = c(\"#fcbf49\", \"#d62828\"), name=\"Haplotype\",\n",
    "                      breaks=c(\"0\",\"1\"),labels=c(\"REF\",\"ALT\")) +\n",
    "    theme(\n",
    "        axis.text.x = element_text(size=16),\n",
    "        axis.title = element_text(size=20),\n",
    "        axis.text.y = element_text(angle = 90),\n",
    "        axis.ticks.x = element_line(),\n",
    "        legend.position = \"top\"\n",
    "      ) +\n",
    "    transp_theme\n",
    "Chr05_Occ_urb_ur_haps\n",
    "ggsave(filename = snakemake@output[[\"Chr05_Occ_urb_ur_haps\"]], plot = Chr05_Occ_urb_ur_haps, \n",
    "       height = 8, width = 20, device = \"pdf\", dpi = 600, units = \"in\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b11d82de-aebd-466c-a996-6b6b215ba13e",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_vcfs <- all_outlier_regions %>% \n",
    "    group_split(start) %>% \n",
    "    purrr::map(., subset_vcf, vcf = chr05_Occ_vcf_urb_sub)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "33679daa-2fdf-45de-b25e-87440480e84c",
   "metadata": {},
   "outputs": [],
   "source": [
    "combined_vcf <- do.call(\"rbind\", all_vcfs)\n",
    "chr05_Occ_urb_sel_AF <- genotype_plot(vcf_object = combined_vcf, popmap = popmap, snp_label_size = 200000, \n",
    "                                      plot_allele_frequency = T)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8cbce361-6d5d-4b74-a29f-2d8d17d20915",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 8, repr.plot.height = 6)\n",
    "cols_hab <- c(\"#007243\", \"#003876\")\n",
    "Chr05_Occ_urb_ur_af <- chr05_Occ_urb_sel_AF$genotypes$data %>% \n",
    "    ggplot(aes(x = AF)) +\n",
    "        geom_density(aes(color = pop, fill = pop), alpha = 0.7) +\n",
    "        scale_color_manual(values = cols_hab) +\n",
    "        scale_fill_manual(values = cols_hab) +\n",
    "        ylab(\"Density\") + \n",
    "        xlab(\"Frequency of REF allele\") +\n",
    "        theme_classic() +\n",
    "        theme(axis.text = element_text(size = 18),\n",
    "              axis.title = element_text(size = 20),\n",
    "              legend.position = 'top',\n",
    "              legend.title = element_text(size = 16),\n",
    "              legend.text = element_text(size = 14)) +\n",
    "        transp_theme\n",
    "Chr05_Occ_urb_ur_af\n",
    "ggsave(filename = snakemake@output[[\"Chr05_Occ_urb_ur_af\"]], plot = Chr05_Occ_urb_ur_af, \n",
    "       height = 6, width = 8, device = \"pdf\", dpi = 600, units = \"in\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "05f19b20-1828-4043-b9dc-29f55e990cde",
   "metadata": {},
   "outputs": [],
   "source": [
    "chr05_Occ_urb_sel_clust <- genotype_plot(vcf_object = combined_vcf, popmap = popmap, snp_label_size = 500000, \n",
    "                                         cluster = T, plot_phased = T)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a28f3ffa-bf09-4b71-988b-6627879d58bd",
   "metadata": {},
   "outputs": [],
   "source": [
    "var_PC1 <- round(get_eigenvalue(chr05_Occ_urb_sel_clust$cluster_pca)[\"Dim.1\",\"variance.percent\"], 1)\n",
    "var_PC2 <- round(get_eigenvalue(chr05_Occ_urb_sel_clust$cluster_pca)[\"Dim.2\",\"variance.percent\"], 1)\n",
    "\n",
    "Chr05_Occ_urb_ur_pca <- chr05_Occ_urb_sel_clust$cluster_pca$l1 %>% \n",
    "    as.data.frame() %>% \n",
    "    rownames_to_column(., var = \"tmp\") %>% \n",
    "    mutate(ind = str_sub(tmp, 1, -3),\n",
    "           hap = str_sub(tmp, -1)) %>% \n",
    "    dplyr::select(ind, hap, RS1, RS2) %>% \n",
    "    left_join(popmap, by = \"ind\") %>% \n",
    "    ggplot(., aes(x = RS1, y = RS2)) +\n",
    "            geom_point(aes(color = pop, shape = pop), size = 7, alpha = 0.75) +\n",
    "            scale_color_manual(values = cols_hab) +\n",
    "            theme_classic() +\n",
    "            xlab(paste0('PC1 (', var_PC1, '%)')) + ylab(paste0('PC2 (', var_PC2, '%)')) +\n",
    "            theme(axis.text = element_text(size = 18),\n",
    "                  axis.title = element_text(size = 20),\n",
    "                  legend.position = 'top',\n",
    "                  legend.title = element_text(size = 16),\n",
    "                  legend.text = element_text(size = 14)) +\n",
    "    transp_theme\n",
    "Chr05_Occ_urb_ur_pca\n",
    "ggsave(filename = snakemake@output[[\"Chr05_Occ_urb_ur_pca\"]], plot = Chr05_Occ_urb_ur_pca, \n",
    "       height = 6, width = 8, device = \"pdf\", dpi = 600, units = \"in\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cef99005-2081-4490-84a1-00c8848e3640",
   "metadata": {},
   "source": [
    "### Positive selection in rural habitats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4a8d9c0c-eee6-41b6-abf8-9c57167f1596",
   "metadata": {},
   "outputs": [],
   "source": [
    "xpnsl_topHits_urban_rural_with_other_stats %>% \n",
    "    filter(direction == 'Rural') %>% \n",
    "    dplyr::select(Chr, start, end, win_size, min_xpnsl_rank, num_nsl_urban:num_fst)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c6f5ac0b-2796-4ca4-bb68-9597039ea6e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "min_xpnsl <- xpnsl_persite %>% filter(normxpnsl == min(normxpnsl))\n",
    "min_xpnsl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b3a57a52-235f-4a06-959c-f4a305cae334",
   "metadata": {},
   "outputs": [],
   "source": [
    "min <- min_xpnsl %>% dplyr::select(Chr, pos) %>% rename('start' = 'pos') %>% mutate(end = start) %>% dplyr::select(Chr, start, end)\n",
    "bt.intersect(min, genes, wb=T)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b02cf4b1-e9e4-4b5f-9572-b42bb39d6726",
   "metadata": {},
   "source": [
    "#### Chr08_Pall"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ddf8a782-62af-4a04-9e28-d186f9ddf3dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "region_df <- xpnsl_topHits_urban_rural %>% \n",
    "    filter(direction == 'Rural') %>% \n",
    "    filter(Chr == 16) %>%\n",
    "    dplyr::select(Chr, start, end)\n",
    "\n",
    "chrom <- region_df %>% pull(Chr) %>% unique()\n",
    "min_pos <- region_df %>% pull(start) %>% min()\n",
    "max_pos <- region_df %>% pull(end) %>% max()\n",
    "buffer <- 500000\n",
    "\n",
    "all_outlier_regions <- xpnsl_rural_outliers %>% \n",
    "    filter(Chr == chrom & start >= min_pos - buffer & end <= max_pos + buffer) %>% \n",
    "    dplyr::select(Chr, start, end)\n",
    "\n",
    "region_raw_df <- xpnsl_persite %>%\n",
    "    filter(Chr == chrom & pos >= min_pos - buffer & pos <= max_pos + buffer)\n",
    "\n",
    "xpnsl_not_outlier <- region_raw_df %>% filter(crit == 0)\n",
    "urban_xpnsl_outliers <- region_raw_df %>% filter(crit == 1 & normxpnsl > 0)\n",
    "rural_xpnsl_outliers <- region_raw_df %>% filter(crit == 1 & normxpnsl < 0)\n",
    "\n",
    "formatter1e6 <- function(x){ \n",
    "    x <- x / 1e6\n",
    "    return(sprintf(x, fmt = '%#.2f'))\n",
    "}\n",
    "\n",
    "fst_outliers <- gt_fst_df_filt %>% filter(is_outlier == 1) %>% dplyr::select(Chr, start, end)\n",
    "fst_wins <- bt.intersect(fst_outliers, all_outlier_regions, wa = T) %>% \n",
    "    rename('Chr' = 'V1', 'start' = 'V2', 'end' = 'V3')\n",
    "nsl_outliers <- nsl_df_filt %>% filter(is_outlier == 1) %>% dplyr::select(Chr, start, end)\n",
    "nsl_wins <- bt.intersect(nsl_outliers, all_outlier_regions, wa = T) %>% \n",
    "    rename('Chr' = 'V1', 'start' = 'V2', 'end' = 'V3')\n",
    "ihs_outliers <- ihs_df_filt %>% filter(is_outlier == 1) %>% dplyr::select(Chr, start, end)\n",
    "ihs_wins <- bt.intersect(ihs_outliers, all_outlier_regions, wa = T) %>% \n",
    "    rename('Chr' = 'V1', 'start' = 'V2', 'end' = 'V3')\n",
    "\n",
    "manhat_plot <- ggplot(xpnsl_not_outlier, aes(x = pos, y = normxpnsl)) +\n",
    "    geom_rect(data=region_df, aes(xmin=start, xmax=end), ymin=-Inf, ymax=Inf, fill=\"black\", alpha=0.3, inherit.aes = F) +\n",
    "    geom_rect(data=all_outlier_regions, aes(xmin=start, xmax=end), ymin=-Inf, ymax=Inf, fill=\"black\", alpha=0.1, inherit.aes = F) +\n",
    "    geom_point(data = xpnsl_not_outlier, shape = 21, alpha = 0.3, size = 1, , fill = \"black\", color = \"black\") +\n",
    "    geom_point(data = urban_xpnsl_outliers, shape = 21, alpha = 0.3, size = 1, color = '#003876', fill = '#003876',\n",
    "               aes(x = pos, y = normxpnsl)) +\n",
    "    geom_point(data = rural_xpnsl_outliers, shape = 21, alpha = 0.3, size = 1, color = '#007243', fill = '#007243',\n",
    "               aes(x = pos, y = normxpnsl)) +\n",
    "    geom_point(data = min_xpnsl, shape = 23, size = 3.5, fill = 'yellow', alpha = 1) +\n",
    "    geom_hline(yintercept = 2, color = \"grey40\", linetype = \"dashed\") +\n",
    "    geom_hline(yintercept = -2, color = \"grey40\", linetype = \"dashed\") +\n",
    "    xlab('Chromosome 16 (position in Mbp)') + ylab('Normalized XP-nSL') +\n",
    "    coord_cartesian(ylim = c(-9.5, 3.5)) +\n",
    "    geom_segment(data = fst_wins, aes(x = start, xend = end), y = -9, yend = -9, color = '#d62828', linewidth = 1) +\n",
    "    geom_segment(data = nsl_wins, aes(x = start, xend = end), y = -9.25, yend = -9.25, color = '#ffbe0b', linewidth = 1) +\n",
    "    geom_segment(data = ihs_wins, aes(x = start, xend = end), y = -9.5, yend = -9.5, color = '#8338ec', linewidth = 1) +\n",
    "    scale_x_continuous(breaks = seq(min_pos - buffer, max_pos + buffer, 200000), labels = formatter1e6) +\n",
    "    scale_y_continuous(breaks = seq(-9, 3, 2)) +\n",
    "    theme_classic() +\n",
    "    theme(\n",
    "        panel.border = element_blank(),\n",
    "        panel.grid.major.x = element_blank(),\n",
    "        panel.grid.minor.x = element_blank(),\n",
    "        axis.text = element_text(size=16),\n",
    "        axis.title = element_text(size=20),\n",
    "      ) +\n",
    "    transp_theme\n",
    "\n",
    "options(repr.plot.width = 8, repr.plot.height = 6)\n",
    "manhat_plot\n",
    "ggsave(filename = snakemake@output[[\"Chr08_Pall_rur_xpnsl\"]], plot = manhat_plot, \n",
    "       height = 6, width = 8, device = \"pdf\", dpi = 600, units = \"in\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "36fa5018-88c1-4f0e-bfdc-060503cab999",
   "metadata": {},
   "outputs": [],
   "source": [
    "chr08_Pall_vcf <- vcfR::read.vcfR(snakemake@input[[\"vcfs\"]][16])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a2376481-c7cd-49d2-9ae1-17a246826bbe",
   "metadata": {},
   "outputs": [],
   "source": [
    "idx <- which(vcfR::getPOS(chr08_Pall_vcf) >= min_pos - buffer & vcfR::getPOS(chr08_Pall_vcf) <= max_pos + buffer)\n",
    "start_idx <- head(idx, n = 1)\n",
    "end_idx <- tail(idx, n = 1)\n",
    "chr08_Pall_vcf_rur_sub <- chr08_Pall_vcf[start_idx:end_idx]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11714882-b2ea-4d57-ac3d-7c9428863ed9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create genotype plot and cluster genotypes with PCA and hclust\n",
    "chr08_Pall_rur_clustered <- genotype_plot(vcf_object = chr08_Pall_vcf_rur_sub, popmap = popmap, snp_label_size = 500000, cluster = T)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "07305cd6-b507-4702-b6b1-973b6594be95",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Reorder popmap based on clustering. This is only for visualization\n",
    "popmap2 <- data.frame(ind = chr08_Pall_rur_clustered$dendro_labels) %>% \n",
    "    left_join(popmap, by = \"ind\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2cd18964-9769-4197-ac93-2f5110b95e0f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create final haplotype plot with colors. Unclustered and split genotypes into phased haplotypes\n",
    "chr08_Pall_rur_final <- genotype_plot(vcf_object = chr08_Pall_vcf_rur_sub, \n",
    "                                     popmap = popmap2, snp_label_size = 200000, \n",
    "                                     cluster = F, plot_phased = T)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "23d1ffcb-458a-4577-b942-d13b3c32e5a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 20, repr.plot.height = 6)\n",
    "chr08_Pall_rur_ur_haps <- chr08_Pall_rur_final$genotypes + \n",
    "    geom_rect(data=region_df, aes(xmin=start, xmax=end), ymin=0, ymax=164.5, fill=\"black\", alpha=0.3, inherit.aes = F) +\n",
    "    geom_rect(data=all_outlier_regions, aes(xmin=start, xmax=end), ymin=0, ymax=164.5, fill=\"black\", alpha=0.1, inherit.aes = F) +\n",
    "    scale_x_continuous(expand = c(0,0), limit = c(min_pos - buffer, max_pos + buffer), \n",
    "                       breaks = seq(min_pos - buffer, max_pos + buffer, 200000), labels = formatter1e6) +\n",
    "    scale_fill_manual(values = c(\"#fcbf49\", \"#d62828\"), name=\"Haplotype\",\n",
    "                      breaks=c(\"0\",\"1\"),labels=c(\"REF\",\"ALT\")) +\n",
    "    theme(\n",
    "        axis.text.x = element_text(size=16),\n",
    "        axis.title = element_text(size=20),\n",
    "        axis.text.y = element_text(angle = 90),\n",
    "        axis.ticks.x = element_line(),\n",
    "        legend.position = \"top\"\n",
    "      ) +\n",
    "    transp_theme\n",
    "chr08_Pall_rur_ur_haps\n",
    "ggsave(filename = snakemake@output[[\"Chr08_Pall_rur_ur_haps\"]], plot = chr08_Pall_rur_ur_haps, \n",
    "       height = 8, width = 20, device = \"pdf\", dpi = 600, units = \"in\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "186cc99b-92c8-4945-8523-85963451a484",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_vcfs <- all_outlier_regions %>% \n",
    "    group_split(start) %>% \n",
    "    purrr::map(., subset_vcf, vcf = chr08_Pall_vcf_rur_sub)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7174ebf2-cff0-4d1f-890c-ff0df58f2d99",
   "metadata": {},
   "outputs": [],
   "source": [
    "combined_vcf <- do.call(\"rbind\", all_vcfs)\n",
    "chr08_Pall_vcf_sel_AF <- genotype_plot(vcf_object = combined_vcf, popmap = popmap, snp_label_size = 200000, \n",
    "                                      plot_allele_frequency = T)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "63df67fc-02ee-4e4e-8556-a6ade271b5ed",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 8, repr.plot.height = 6)\n",
    "cols_hab <- c(\"#007243\", \"#003876\")\n",
    "chr08_Pall_rur_ur_af <- chr08_Pall_vcf_sel_AF$genotypes$data %>% \n",
    "    ggplot(aes(x = AF)) +\n",
    "        geom_density(aes(color = pop, fill = pop), alpha = 0.7) +\n",
    "        scale_color_manual(values = cols_hab) +\n",
    "        scale_fill_manual(values = cols_hab) +\n",
    "        ylab(\"Density\") + \n",
    "        xlab(\"Frequency of REF allele\") +\n",
    "        theme_classic() +\n",
    "        theme(axis.text = element_text(size = 18),\n",
    "              axis.title = element_text(size = 20),\n",
    "              legend.position = 'top',\n",
    "              legend.title = element_text(size = 16),\n",
    "              legend.text = element_text(size = 14)) +\n",
    "        transp_theme\n",
    "chr08_Pall_rur_ur_af\n",
    "ggsave(filename = snakemake@output[[\"Chr08_Pall_rur_ur_af\"]], plot = chr08_Pall_rur_ur_af, \n",
    "       height = 6, width = 8, device = \"pdf\", dpi = 600, units = \"in\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "09c1fef8-6751-4a04-b739-8a7fab5fae33",
   "metadata": {},
   "outputs": [],
   "source": [
    "chr08_Pall_rur_sel_clust <- genotype_plot(vcf_object = combined_vcf, popmap = popmap, snp_label_size = 500000, \n",
    "                                         cluster = T, plot_phased = T)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b342df55-8c73-401a-a095-e11fa3591323",
   "metadata": {},
   "outputs": [],
   "source": [
    "var_PC1 <- round(get_eigenvalue(chr08_Pall_rur_sel_clust$cluster_pca)[\"Dim.1\",\"variance.percent\"], 1)\n",
    "var_PC2 <- round(get_eigenvalue(chr08_Pall_rur_sel_clust$cluster_pca)[\"Dim.2\",\"variance.percent\"], 1)\n",
    "\n",
    "chr08_Pall_rur_ur_pca <- chr08_Pall_rur_sel_clust$cluster_pca$l1 %>% \n",
    "    as.data.frame() %>% \n",
    "    rownames_to_column(., var = \"tmp\") %>% \n",
    "    mutate(ind = str_sub(tmp, 1, -3),\n",
    "           hap = str_sub(tmp, -1)) %>% \n",
    "    dplyr::select(ind, hap, RS1, RS2) %>% \n",
    "    left_join(popmap, by = \"ind\") %>% \n",
    "    ggplot(., aes(x = RS1, y = RS2)) +\n",
    "            geom_point(aes(color = pop, shape = pop), size = 7, alpha = 0.75) +\n",
    "            scale_color_manual(values = cols_hab) +\n",
    "            theme_classic() +\n",
    "            xlab(paste0('PC1 (', var_PC1, '%)')) + ylab(paste0('PC2 (', var_PC2, '%)')) +\n",
    "            theme(axis.text = element_text(size = 18),\n",
    "                  axis.title = element_text(size = 20),\n",
    "                  legend.position = 'top',\n",
    "                  legend.title = element_text(size = 16),\n",
    "                  legend.text = element_text(size = 14)) +\n",
    "    transp_theme\n",
    "chr08_Pall_rur_ur_pca\n",
    "ggsave(filename = snakemake@output[[\"Chr08_Pall_rur_ur_pca\"]], plot = chr08_Pall_rur_ur_pca, \n",
    "       height = 6, width = 8, device = \"pdf\", dpi = 600, units = \"in\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b6360ccc-ad3e-475b-8d36-0d91ec29476a",
   "metadata": {},
   "source": [
    "#### Chr04_Occ"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f8e28e2c-e474-40bb-a246-7eea0c6faa3e",
   "metadata": {},
   "outputs": [],
   "source": [
    "region_df <- xpnsl_topHits_urban_rural %>% \n",
    "    filter(direction == 'Rural') %>% \n",
    "    filter(Chr == 7) %>%\n",
    "    dplyr::select(Chr, start, end)\n",
    "\n",
    "chrom <- region_df %>% pull(Chr) %>% unique()\n",
    "min_pos <- region_df %>% pull(start) %>% min()\n",
    "max_pos <- region_df %>% pull(end) %>% max()\n",
    "buffer <- 500000\n",
    "\n",
    "all_outlier_regions <- xpnsl_rural_outliers %>% \n",
    "    filter(Chr == chrom & start >= min_pos - buffer & end <= max_pos + buffer) %>% \n",
    "    dplyr::select(Chr, start, end)\n",
    "\n",
    "region_raw_df <- xpnsl_persite %>%\n",
    "    filter(Chr == chrom & pos >= min_pos - buffer & pos <= max_pos + buffer)\n",
    "\n",
    "xpnsl_not_outlier <- region_raw_df %>% filter(crit == 0)\n",
    "urban_xpnsl_outliers <- region_raw_df %>% filter(crit == 1 & normxpnsl > 0)\n",
    "rural_xpnsl_outliers <- region_raw_df %>% filter(crit == 1 & normxpnsl < 0)\n",
    "\n",
    "formatter1e6 <- function(x){ \n",
    "    x <- x / 1e6\n",
    "    return(sprintf(x, fmt = '%#.2f'))\n",
    "}\n",
    "fst_outliers <- gt_fst_df_filt %>% filter(is_outlier == 1) %>% dplyr::select(Chr, start, end)\n",
    "fst_wins <- bt.intersect(fst_outliers, all_outlier_regions, wa = T) %>% \n",
    "    rename('Chr' = 'V1', 'start' = 'V2', 'end' = 'V3')\n",
    "nsl_outliers <- nsl_df_filt %>% filter(is_outlier == 1 & habitat == \"Urban\") %>% dplyr::select(Chr, start, end)\n",
    "nsl_wins <- bt.intersect(nsl_outliers, all_outlier_regions, wa = T) %>% \n",
    "    rename('Chr' = 'V1', 'start' = 'V2', 'end' = 'V3')\n",
    "ihs_outliers <- ihs_df_filt %>% filter(is_outlier == 1 & habitat == \"Urban\") %>% dplyr::select(Chr, start, end)\n",
    "ihs_wins <- bt.intersect(ihs_outliers, all_outlier_regions, wa = T) %>% \n",
    "    rename('Chr' = 'V1', 'start' = 'V2', 'end' = 'V3')\n",
    "\n",
    "manhat_plot <- ggplot(xpnsl_not_outlier, aes(x = pos, y = normxpnsl)) +\n",
    "    geom_rect(data=region_df, aes(xmin=start, xmax=end), ymin=-Inf, ymax=Inf, fill=\"black\", alpha=0.3, inherit.aes = F) +\n",
    "    geom_rect(data=all_outlier_regions, aes(xmin=start, xmax=end), ymin=-Inf, ymax=Inf, fill=\"black\", alpha=0.1, inherit.aes = F) +\n",
    "    geom_point(data = xpnsl_not_outlier, shape = 21, alpha = 0.3, size = 1, , fill = \"black\", color = \"black\") +\n",
    "    geom_point(data = urban_xpnsl_outliers, shape = 21, alpha = 0.3, size = 1, color = '#003876', fill = '#003876',\n",
    "               aes(x = pos, y = normxpnsl)) +\n",
    "    geom_point(data = rural_xpnsl_outliers, shape = 21, alpha = 0.3, size = 1, color = '#007243', fill = '#007243',\n",
    "               aes(x = pos, y = normxpnsl)) +\n",
    "    geom_hline(yintercept = 2, color = \"grey40\", linetype = \"dashed\") +\n",
    "    geom_hline(yintercept = -2, color = \"grey40\", linetype = \"dashed\") +\n",
    "    xlab('Chromosome 7 (position in Mbp)') + ylab('Normalized XP-nSL') +\n",
    "    coord_cartesian(ylim = c(-6.5, 3.5)) +\n",
    "    scale_x_continuous(breaks = seq(min_pos - buffer, max_pos + buffer, 200000), labels = formatter1e6) +\n",
    "    geom_segment(data = fst_wins, aes(x = start, xend = end), y = -6, yend = -6, color = '#d62828', linewidth = 1) +\n",
    "    geom_segment(data = nsl_wins, aes(x = start, xend = end), y = -6.25, yend = -6.25, color = '#ffbe0b', linewidth = 1) +\n",
    "    geom_segment(data = ihs_wins, aes(x = start, xend = end), y = -6.5, yend = -6.5, color = '#8338ec', linewidth = 1) +\n",
    "    scale_y_continuous(breaks = seq(-6, 4, 2)) +\n",
    "    theme_classic() +\n",
    "    theme(\n",
    "        panel.border = element_blank(),\n",
    "        panel.grid.major.x = element_blank(),\n",
    "        panel.grid.minor.x = element_blank(),\n",
    "        axis.text = element_text(size=16),\n",
    "        axis.title = element_text(size=20),\n",
    "      ) +\n",
    "    transp_theme\n",
    "\n",
    "options(repr.plot.width = 8, repr.plot.height = 6)\n",
    "manhat_plot\n",
    "ggsave(filename = snakemake@output[[\"Chr04_Occ_rur_xpnsl\"]], plot = manhat_plot, \n",
    "       height = 6, width = 8, device = \"pdf\", dpi = 600, units = \"in\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9c3b988a-3c34-490f-9195-47d457b99e34",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Subset VCF around same region as above\n",
    "idx <- which(vcfR::getPOS(chr04_Occ_vcf) >= min_pos - buffer & vcfR::getPOS(chr04_Occ_vcf) <= max_pos + buffer)\n",
    "start_idx <- head(idx, n = 1)\n",
    "end_idx <- tail(idx, n = 1)\n",
    "chr04_Occ_vcf_rur_sub <- chr04_Occ_vcf[start_idx:end_idx]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e484b5b1-0f0d-4486-a2eb-72bff1d8ba29",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create genotype plot and cluster genotypes with PCA and hclust\n",
    "chr04_Occ_rur_clustered <- genotype_plot(vcf_object = chr04_Occ_vcf_rur_sub, popmap = popmap, snp_label_size = 200000, cluster = T)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "759481e9-3a59-49ad-9519-4b22b845d25c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Reorder popmap based on clustering. This is only for visualization\n",
    "popmap2 <- data.frame(ind = chr04_Occ_rur_clustered$dendro_labels) %>% \n",
    "    left_join(popmap, by = \"ind\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a3b6e09b-278f-4569-a2e9-464e80df857e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create final haplotype plot with colors. Unclustered and split genotypes into phased haplotypes\n",
    "chr04_Occ_rur_final <- genotype_plot(vcf_object = chr04_Occ_vcf_rur_sub, \n",
    "                                     popmap = popmap2, snp_label_size = 200000, \n",
    "                                     cluster = F, plot_phased = T)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c4f029d2-2507-4db2-a224-96075e270171",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 20, repr.plot.height = 6)\n",
    "Chr04_Occ_rur_ur_haps <- chr04_Occ_rur_final$genotypes + \n",
    "    geom_rect(data=region_df, aes(xmin=start, xmax=end), ymin=0, ymax=164.5, fill=\"black\", alpha=0.3, inherit.aes = F) +\n",
    "    geom_rect(data=all_outlier_regions, aes(xmin=start, xmax=end), ymin=0, ymax=164.5, fill=\"black\", alpha=0.1, inherit.aes = F) +\n",
    "    scale_x_continuous(expand = c(0,0), limit = c(min_pos - buffer, max_pos + buffer), \n",
    "                       breaks = seq(min_pos - buffer, max_pos + buffer, 200000), labels = formatter1e6) +\n",
    "    scale_fill_manual(values = c(\"#fcbf49\", \"#d62828\"), name=\"Haplotype\",\n",
    "                      breaks=c(\"0\",\"1\"),labels=c(\"REF\",\"ALT\")) +\n",
    "    theme(\n",
    "        axis.text.x = element_text(size=16),\n",
    "        axis.title = element_text(size=20),\n",
    "        axis.text.y = element_text(angle = 90),\n",
    "        axis.ticks.x = element_line(),\n",
    "        legend.position = \"top\"\n",
    "      ) +\n",
    "    transp_theme\n",
    "Chr04_Occ_rur_ur_haps\n",
    "ggsave(filename = snakemake@output[[\"Chr04_Occ_rur_ur_haps\"]], plot = Chr04_Occ_rur_ur_haps, \n",
    "       height = 8, width = 20, device = \"pdf\", dpi = 600, units = \"in\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4dbcee0b-0629-41fd-8fdb-4751793fe4bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_vcfs <- all_outlier_regions %>% \n",
    "    group_split(start) %>% \n",
    "    purrr::map(., subset_vcf, vcf = chr04_Occ_vcf_rur_sub)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0e881dcc-d180-41c6-9b0f-7700df447b62",
   "metadata": {},
   "outputs": [],
   "source": [
    "combined_vcf <- do.call(\"rbind\", all_vcfs)\n",
    "chr04_Occ_rur_sel_AF <- genotype_plot(vcf_object = combined_vcf, popmap = popmap, snp_label_size = 200000, \n",
    "                                      plot_allele_frequency = T)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "38087b9e-1a5c-4f90-8694-e9181f34ca77",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 8, repr.plot.height = 6)\n",
    "cols_hab <- c(\"#007243\", \"#003876\")\n",
    "Chr04_Occ_rur_ur_af <- chr04_Occ_rur_sel_AF$genotypes$data %>% \n",
    "    ggplot(aes(x = AF)) +\n",
    "        geom_density(aes(color = pop, fill = pop), alpha = 0.7) +\n",
    "        scale_color_manual(values = cols_hab) +\n",
    "        scale_fill_manual(values = cols_hab) +\n",
    "        ylab(\"Density\") + \n",
    "        xlab(\"Frequency of REF allele\") +\n",
    "        theme_classic() +\n",
    "        theme(axis.text = element_text(size = 18),\n",
    "              axis.title = element_text(size = 20),\n",
    "              legend.position = 'top',\n",
    "              legend.title = element_text(size = 16),\n",
    "              legend.text = element_text(size = 14)) +\n",
    "        transp_theme\n",
    "Chr04_Occ_rur_ur_af\n",
    "ggsave(filename = snakemake@output[[\"Chr04_Occ_rur_ur_af\"]], plot = Chr04_Occ_rur_ur_af, \n",
    "       height = 6, width = 8, device = \"pdf\", dpi = 600, units = \"in\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f2aeabb4-0848-4a46-ad7e-29d5653caeac",
   "metadata": {},
   "outputs": [],
   "source": [
    "chr04_Occ_rur_sel_clust <- genotype_plot(vcf_object = combined_vcf, popmap = popmap, snp_label_size = 200000, \n",
    "                                         cluster = T, plot_phased = T)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a4554e8e-d9c2-42ff-a8b9-00129720a1f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "var_PC1 <- round(get_eigenvalue(chr04_Occ_rur_sel_clust$cluster_pca)[\"Dim.1\",\"variance.percent\"], 1)\n",
    "var_PC2 <- round(get_eigenvalue(chr04_Occ_rur_sel_clust$cluster_pca)[\"Dim.2\",\"variance.percent\"], 1)\n",
    "\n",
    "Chr04_Occ_rur_ur_pca <- chr04_Occ_rur_sel_clust$cluster_pca$l1 %>% \n",
    "    as.data.frame() %>% \n",
    "    rownames_to_column(., var = \"tmp\") %>% \n",
    "    mutate(ind = str_sub(tmp, 1, -3),\n",
    "           hap = str_sub(tmp, -1)) %>% \n",
    "    dplyr::select(ind, hap, RS1, RS2) %>% \n",
    "    left_join(popmap, by = \"ind\") %>% \n",
    "    ggplot(., aes(x = RS1, y = RS2)) +\n",
    "            geom_point(aes(color = pop, shape = pop), size = 7, alpha = 0.75) +\n",
    "            scale_color_manual(values = cols_hab) +\n",
    "            theme_classic() +\n",
    "            xlab(paste0('PC1 (', var_PC1, '%)')) + ylab(paste0('PC2 (', var_PC2, '%)')) +\n",
    "            theme(axis.text = element_text(size = 18),\n",
    "                  axis.title = element_text(size = 20),\n",
    "                  legend.position = 'top',\n",
    "                  legend.title = element_text(size = 16),\n",
    "                  legend.text = element_text(size = 14)) +\n",
    "    transp_theme\n",
    "Chr04_Occ_rur_ur_pca\n",
    "ggsave(filename = snakemake@output[[\"Chr04_Occ_rur_ur_pca\"]], plot = Chr04_Occ_rur_ur_pca, \n",
    "       height = 6, width = 8, device = \"pdf\", dpi = 600, units = \"in\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d02ba4a5-6536-4116-9abb-3dd6f685c65c",
   "metadata": {},
   "source": [
    "## Haplotype PCA and AF for random unselected regions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bec22e11-3220-4fdb-97fc-c10ecb44b201",
   "metadata": {},
   "outputs": [],
   "source": [
    "set.seed(42)\n",
    "random_unsel_regions <- obs_xpnsl_ur_df_filt %>% \n",
    "    filter(direction == \"None\" & Chr == 1) %>% \n",
    "    dplyr::select(Chr, start, end) %>% \n",
    "    sample_n(100) %>% \n",
    "    arrange(Chr, start)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "51fceeb7-0753-4f94-ad72-5b24125081aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "random_unsel_regions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8b13cea5-401f-4774-bd9e-e1f04d67ab91",
   "metadata": {},
   "outputs": [],
   "source": [
    "subset_vcf <- function(df){\n",
    "    chrom = \"Chr01_Occ\"\n",
    "    vcf_idx <- grep(chrom, snakemake@input[[\"vcfs\"]])\n",
    "    vcf <- vcfR::read.vcfR(snakemake@input[[\"vcfs\"]][vcf_idx])\n",
    "\n",
    "    starts <- df$start\n",
    "    ends <- df$end\n",
    "\n",
    "    sub_vcfs <- list()\n",
    "    for(i in 1:length(starts)){\n",
    "        idx <- which(vcfR::getPOS(vcf) >= starts[i] & vcfR::getPOS(vcf) <= ends[i])\n",
    "        start_idx <- head(idx, n = 1)\n",
    "        end_idx <- tail(idx, n = 1)\n",
    "        vcf_sub <- vcf[start_idx:end_idx]\n",
    "        sub_vcfs[[i]] <- vcf_sub\n",
    "    }\n",
    "    comb_vcf <- do.call(rbind, sub_vcfs)\n",
    "    \n",
    "    return(comb_vcf)\n",
    "}\n",
    "\n",
    "random_unsel_regions_vcf_list <- random_unsel_regions %>% \n",
    "    group_split(Chr) %>% \n",
    "    purrr::map(., subset_vcf)\n",
    "random_unsel_regions_vcf <- do.call(rbind, random_unsel_regions_vcf_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a0717f20-ceec-419f-99d2-2816b53dce76",
   "metadata": {},
   "outputs": [],
   "source": [
    "random_unsel_regions_vcf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "45b21458-e610-4256-b21e-3357099072ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "random_unsel_regions_AF <- genotype_plot(vcf_object = random_unsel_regions_vcf, popmap = popmap, snp_label_size = 500000, \n",
    "                                         plot_allele_frequency = T)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a59f52bc-3b37-4ea1-ad09-a6676fd0b234",
   "metadata": {},
   "outputs": [],
   "source": [
    "cols_hab <- c(\"#007243\", \"#003876\")\n",
    "\n",
    "random_unsel_regions_af <- random_unsel_regions_AF$genotypes$data %>% \n",
    "    ggplot(aes(x = AF)) +\n",
    "        geom_density(aes(color = pop, fill = pop), alpha = 0.7) +\n",
    "        scale_color_manual(values = cols_hab) +\n",
    "        scale_fill_manual(values = cols_hab) +\n",
    "        ylab(\"Density\") + \n",
    "        xlab(\"Frequency of REF allele\") +\n",
    "        theme_classic() +\n",
    "        theme(axis.text = element_text(size = 18),\n",
    "              axis.title = element_text(size = 20),\n",
    "              legend.position = 'top',\n",
    "              legend.title = element_text(size = 16),\n",
    "              legend.text = element_text(size = 14)) +\n",
    "        transp_theme\n",
    "random_unsel_regions_af\n",
    "\n",
    "ggsave(filename = snakemake@output[[\"random_unsel_regions_af\"]], plot = random_unsel_regions_af, \n",
    "       height = 6, width = 8, device = \"pdf\", dpi = 600, units = \"in\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5889efa8-21a4-4b02-b82d-2a22ed96ae13",
   "metadata": {},
   "outputs": [],
   "source": [
    "random_unsel_regions_clust <- genotype_plot(vcf_object = random_unsel_regions_vcf, popmap = popmap, snp_label_size = 200000, \n",
    "                                            cluster = T, plot_phased = T)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b3fc7ccd-7a6d-45f9-9cd1-80c8d90a27c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "var_PC1 <- round(get_eigenvalue(random_unsel_regions_clust$cluster_pca)[\"Dim.1\",\"variance.percent\"], 1)\n",
    "var_PC2 <- round(get_eigenvalue(random_unsel_regions_clust$cluster_pca)[\"Dim.2\",\"variance.percent\"], 1)\n",
    "\n",
    "random_unsel_regions_pca <- random_unsel_regions_clust$cluster_pca$l1 %>% \n",
    "    as.data.frame() %>% \n",
    "    rownames_to_column(., var = \"tmp\") %>% \n",
    "    mutate(ind = str_sub(tmp, 1, -3),\n",
    "           hap = str_sub(tmp, -1)) %>% \n",
    "    dplyr::select(ind, hap, RS1, RS2) %>% \n",
    "    left_join(popmap, by = \"ind\") %>% \n",
    "    ggplot(., aes(x = RS1, y = RS2)) +\n",
    "            geom_point(aes(color = pop, shape = pop), size = 7, alpha = 0.75) +\n",
    "            scale_color_manual(values = cols_hab) +\n",
    "            theme_classic() +\n",
    "            xlab(paste0('PC1 (', var_PC1, '%)')) + ylab(paste0('PC2 (', var_PC2, '%)')) +\n",
    "            theme(axis.text = element_text(size = 18),\n",
    "                  axis.title = element_text(size = 20),\n",
    "                  legend.position = 'top',\n",
    "                  legend.title = element_text(size = 16),\n",
    "                  legend.text = element_text(size = 14)) +\n",
    "    transp_theme\n",
    "random_unsel_regions_pca\n",
    "\n",
    "ggsave(filename = snakemake@output[[\"random_unsel_regions_pca\"]], plot = random_unsel_regions_pca, \n",
    "       height = 6, width = 8, device = \"pdf\", dpi = 600, units = \"in\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
